{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f2e25f8e-3183-421d-9a3b-2adc6c958c41",
   "metadata": {},
   "source": [
    "# MEPP Walkthrough: Motif Enrichment Positional Profiling\n",
    "This notebook will walk you through how to perform a score based motif enrichment analysis that profiles motif enrichment at multiple positions across a set of sequences centered on biologically relevant features, in this case a central binding motif of interest."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84c79a9a-0b01-4e9d-90a1-e31210cbf964",
   "metadata": {},
   "source": [
    "# Install prerequisites\n",
    "\n",
    "You will need the following prerequisites:\n",
    "* MEPP\n",
    "* HOMER\n",
    "* pandas\n",
    "* numpy\n",
    "* coolbox\n",
    "* wget\n",
    "* samtools\n",
    "* deeptools\n",
    "* bedtools\n",
    "* bedops\n",
    "* wiggletools\n",
    "* wigToBigWig\n",
    "* tensorflow\n",
    "\n",
    "To install most of these through mamba:\n",
    "```\n",
    "mamba create -d -n mepp_walkthrough -c bioconda -c conda-forge homer pandas numpy coolbox wget samtools deeptools bedtools bedops wiggletools ucsc-wigtobigwig tensorflow\n",
    "```\n",
    "\n",
    "To install most of these through conda (slower):\n",
    "```\n",
    "conda create -d -n mepp_walkthrough -c bioconda -c conda-forge homer pandas numpy coolbox wget samtools deeptools bedtools bedops wiggletools ucsc-wigtobigwig tensorflow\n",
    "```\n",
    "\n",
    "To activate the environment:\n",
    "```\n",
    "conda activate mepp_walkthrough\n",
    "```\n",
    "\n",
    "To install MEPP, use pip:\n",
    "```\n",
    "pip install git+https://github.com/npdeloss/mepp@main\n",
    "```\n",
    "\n",
    "Or, if you only have user privileges:\n",
    "```\n",
    "pip install git+https://github.com/npdeloss/mepp@main --user\n",
    "```\n",
    "\n",
    "You may need to append the following to your ~/.bashrc:\n",
    "```\n",
    "export PATH=\"$HOME/.local/bin:$PATH\"\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "201cf6d2-a242-4823-a107-3a0657dc70ea",
   "metadata": {},
   "source": [
    "# Import key libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ebd4344f-c08c-4773-b965-8de1f819bb51",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55eed5b8-ff4c-4ffd-979c-33d82feee37a",
   "metadata": {},
   "source": [
    "# Enumerate sample sheet with alignment files for download\n",
    "Here we will be comparing K562 and HCT116 cell lines from ENCODE."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "667127b4-1d26-4edc-9b49-5bdd3701d17e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing k562.vs.hct116.atac-seq.samples.txt\n"
     ]
    }
   ],
   "source": [
    "%%file k562.vs.hct116.atac-seq.samples.txt\n",
    "cell_type replicate bam_url\n",
    "k562 1 https://www.encodeproject.org/files/ENCFF512VEZ/@@download/ENCFF512VEZ.bam\n",
    "k562 2 https://www.encodeproject.org/files/ENCFF987XOV/@@download/ENCFF987XOV.bam\n",
    "hct116 1 https://www.encodeproject.org/files/ENCFF724QHH/@@download/ENCFF724QHH.bam\n",
    "hct116 2 https://www.encodeproject.org/files/ENCFF927YUB/@@download/ENCFF927YUB.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03573cae-4553-4d0b-8b5f-e320a939d806",
   "metadata": {},
   "source": [
    "# Load sample sheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ee247ff7-129b-436f-a332-5771d6073c41",
   "metadata": {},
   "outputs": [],
   "source": [
    "samplesheet_filepath = 'k562.vs.hct116.atac-seq.samples.txt'\n",
    "samplesheet_sep = ' '"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "93ed8321-5a32-4f85-8cd0-f9cf7e1daf7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_type</th>\n",
       "      <th>replicate</th>\n",
       "      <th>bam_url</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k562</td>\n",
       "      <td>1</td>\n",
       "      <td>https://www.encodeproject.org/files/ENCFF512VE...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k562</td>\n",
       "      <td>2</td>\n",
       "      <td>https://www.encodeproject.org/files/ENCFF987XO...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hct116</td>\n",
       "      <td>1</td>\n",
       "      <td>https://www.encodeproject.org/files/ENCFF724QH...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hct116</td>\n",
       "      <td>2</td>\n",
       "      <td>https://www.encodeproject.org/files/ENCFF927YU...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  cell_type  replicate                                            bam_url  \\\n",
       "0      k562          1  https://www.encodeproject.org/files/ENCFF512VE...   \n",
       "1      k562          2  https://www.encodeproject.org/files/ENCFF987XO...   \n",
       "2    hct116          1  https://www.encodeproject.org/files/ENCFF724QH...   \n",
       "3    hct116          2  https://www.encodeproject.org/files/ENCFF927YU...   \n",
       "\n",
       "   sample  \n",
       "0    True  \n",
       "1    True  \n",
       "2    True  \n",
       "3    True  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samplesheet_df = pd.read_csv(samplesheet_filepath, sep = samplesheet_sep)\n",
    "samplesheet_df['sample'] = True\n",
    "samplesheet_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8270b032-1097-4a58-9cb3-c72228a1f2c1",
   "metadata": {},
   "source": [
    "# Download Alignment files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5b13a1cb-e451-4e7d-8edd-029258c2e001",
   "metadata": {},
   "outputs": [],
   "source": [
    "samplesheet_df['basename'] = samplesheet_df['cell_type'] + '_rep' + samplesheet_df['replicate'].astype(str)\n",
    "samplesheet_df['bam_filepath'] = samplesheet_df['basename'] + '.bam'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f559c495-0125-4a61-8e69-bc4c1f233698",
   "metadata": {},
   "outputs": [],
   "source": [
    "samplesheet_df['wget_bam_cmd'] = (\n",
    "    'wget -nc -O ' +\n",
    "    samplesheet_df['bam_filepath'] + ' ' +\n",
    "    '\"' + samplesheet_df['bam_url'] + '\"'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "30b3d3ab-82e4-45a5-874b-e6457f59afc4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wget -nc -O k562_rep1.bam \"https://www.encodeproject.org/files/ENCFF512VEZ/@@download/ENCFF512VEZ.bam\"\n",
      "--2022-01-27 11:15:29--  https://www.encodeproject.org/files/ENCFF512VEZ/@@download/ENCFF512VEZ.bam\n",
      "Resolving www.encodeproject.org (www.encodeproject.org)... 34.211.244.144\n",
      "Connecting to www.encodeproject.org (www.encodeproject.org)|34.211.244.144|:443... connected.\n",
      "HTTP request sent, awaiting response... 307 Temporary Redirect\n",
      "Location: https://encode-public.s3.amazonaws.com/2021/02/24/c7d946d1-a19f-4c94-988f-35ef5edc67e9/ENCFF512VEZ.bam?response-content-disposition=attachment%3B%20filename%3DENCFF512VEZ.bam&AWSAccessKeyId=ASIATGZNGCNXVI6AN3H7&Signature=oyrrr6phc%2FVKa583BeKDUdNPvtk%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEJv%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJHMEUCIDim9%2FmXf7aVc2BrNnzI1trmTuAwp6%2F9kM5CwIRf0XDiAiEAjZG1lhLINFny2vUpvSVfsA9vbd6MuXY3pGgta5IgbnIqgwQIxP%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FARAAGgwyMjA3NDg3MTQ4NjMiDJMoTf9SoSH9PxnSnyrXAyYAVknoW4G7u04vu50%2Bv6fhmX5wO1KLuZtDdVMfycNGhQ4zcVijYCPHY6ohjKsfNxdTPVS03VcqB3PR3DWWt0LDBgo0cXsL4dVSqmeaB1X5MUojgjoaza9jutFAptk%2BHGcJYaN6WGcYP%2FsP39KrNmhqwIgr8pNASQtBQrSLtTIItR89GUZXfODiHNUM%2FdZPad10A6PFnqVLiZ195hDByN%2FmyL4jIWfgFJE8wJhI1KHm0GY37sAYoxkP7NH3Bfr00QArKL0XB%2BL9i50GQa9lUIpgS8Iq8E7qgD4oJaT4Oe3gbMB9m4mS%2FmSGI%2Bj57o%2FO3Qct98oD%2FyzH9tJff65v8b3OyNnvTQLwLQbCXyCx1EGqhRD%2BHLFQCT%2B5kJGS5O9bKMzvGeMYpcnRVfjjmhBPEQ2v0Q%2F1rxHOhHC5SQapH2rUSVlAwo%2B467VwyanlqQnXkgnjtVkt%2Bw5roWvelWbuDrLypaap9gEXah9PfOS9KbRSCBDZxZKkgtitDqZHd7eCpaW3C20jqDs4rftNfDlZZA%2FlSIgLK46ERrTixrBvqBg2bTf484Zh6SvxG4bDKRdb9R%2FP1XsZP9pBKlLFxW3IjTy%2FeKybXcw3DmkcuM7JbwhAxVb2vD7YDjCSxsuPBjqlAVNDb7aL7uWi9Bz%2BNOKuubbQm%2FQ7ScxfWAqwn2thzH%2BGHkjByv%2BLUg4egwXVcWfK6j1LK4Wu%2FQMu8WLzQwllr2AN9QaOK3S9JXRCsdZMaqGGhMxOgRzd6q8WVNlrUAf8Xfti%2FgyjM3h4jU3Xclzd7ZBKFF8PhuQ6RCdKMQc%2BGW4H0SjWN9YqEPznqPDtchth9%2FAl7M20PDNSGvTUan4PNBDtUQOnag%3D%3D&Expires=1643440529 [following]\n",
      "--2022-01-27 11:15:29--  https://encode-public.s3.amazonaws.com/2021/02/24/c7d946d1-a19f-4c94-988f-35ef5edc67e9/ENCFF512VEZ.bam?response-content-disposition=attachment%3B%20filename%3DENCFF512VEZ.bam&AWSAccessKeyId=ASIATGZNGCNXVI6AN3H7&Signature=oyrrr6phc%2FVKa583BeKDUdNPvtk%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEJv%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJHMEUCIDim9%2FmXf7aVc2BrNnzI1trmTuAwp6%2F9kM5CwIRf0XDiAiEAjZG1lhLINFny2vUpvSVfsA9vbd6MuXY3pGgta5IgbnIqgwQIxP%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FARAAGgwyMjA3NDg3MTQ4NjMiDJMoTf9SoSH9PxnSnyrXAyYAVknoW4G7u04vu50%2Bv6fhmX5wO1KLuZtDdVMfycNGhQ4zcVijYCPHY6ohjKsfNxdTPVS03VcqB3PR3DWWt0LDBgo0cXsL4dVSqmeaB1X5MUojgjoaza9jutFAptk%2BHGcJYaN6WGcYP%2FsP39KrNmhqwIgr8pNASQtBQrSLtTIItR89GUZXfODiHNUM%2FdZPad10A6PFnqVLiZ195hDByN%2FmyL4jIWfgFJE8wJhI1KHm0GY37sAYoxkP7NH3Bfr00QArKL0XB%2BL9i50GQa9lUIpgS8Iq8E7qgD4oJaT4Oe3gbMB9m4mS%2FmSGI%2Bj57o%2FO3Qct98oD%2FyzH9tJff65v8b3OyNnvTQLwLQbCXyCx1EGqhRD%2BHLFQCT%2B5kJGS5O9bKMzvGeMYpcnRVfjjmhBPEQ2v0Q%2F1rxHOhHC5SQapH2rUSVlAwo%2B467VwyanlqQnXkgnjtVkt%2Bw5roWvelWbuDrLypaap9gEXah9PfOS9KbRSCBDZxZKkgtitDqZHd7eCpaW3C20jqDs4rftNfDlZZA%2FlSIgLK46ERrTixrBvqBg2bTf484Zh6SvxG4bDKRdb9R%2FP1XsZP9pBKlLFxW3IjTy%2FeKybXcw3DmkcuM7JbwhAxVb2vD7YDjCSxsuPBjqlAVNDb7aL7uWi9Bz%2BNOKuubbQm%2FQ7ScxfWAqwn2thzH%2BGHkjByv%2BLUg4egwXVcWfK6j1LK4Wu%2FQMu8WLzQwllr2AN9QaOK3S9JXRCsdZMaqGGhMxOgRzd6q8WVNlrUAf8Xfti%2FgyjM3h4jU3Xclzd7ZBKFF8PhuQ6RCdKMQc%2BGW4H0SjWN9YqEPznqPDtchth9%2FAl7M20PDNSGvTUan4PNBDtUQOnag%3D%3D&Expires=1643440529\n",
      "Resolving encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)... 52.218.250.227\n",
      "Connecting to encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)|52.218.250.227|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 3132141788 (2.9G) [binary/octet-stream]\n",
      "Saving to: ‘k562_rep1.bam’\n",
      "\n",
      "k562_rep1.bam       100%[===================>]   2.92G  12.6MB/s    in 3m 45s  \n",
      "\n",
      "2022-01-27 11:19:14 (13.3 MB/s) - ‘k562_rep1.bam’ saved [3132141788/3132141788]\n",
      "\n",
      "wget -nc -O k562_rep2.bam \"https://www.encodeproject.org/files/ENCFF987XOV/@@download/ENCFF987XOV.bam\"\n",
      "--2022-01-27 11:19:14--  https://www.encodeproject.org/files/ENCFF987XOV/@@download/ENCFF987XOV.bam\n",
      "Resolving www.encodeproject.org (www.encodeproject.org)... 34.211.244.144\n",
      "Connecting to www.encodeproject.org (www.encodeproject.org)|34.211.244.144|:443... connected.\n",
      "HTTP request sent, awaiting response... 307 Temporary Redirect\n",
      "Location: https://encode-public.s3.amazonaws.com/2021/02/24/57ce8c5b-7229-44b1-b5dc-e62d82a5b7d2/ENCFF987XOV.bam?response-content-disposition=attachment%3B%20filename%3DENCFF987XOV.bam&AWSAccessKeyId=ASIATGZNGCNXVI6AN3H7&Signature=UVv5jBum8NR4CIy1dN1y2ZUoEhc%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEJv%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJHMEUCIDim9%2FmXf7aVc2BrNnzI1trmTuAwp6%2F9kM5CwIRf0XDiAiEAjZG1lhLINFny2vUpvSVfsA9vbd6MuXY3pGgta5IgbnIqgwQIxP%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FARAAGgwyMjA3NDg3MTQ4NjMiDJMoTf9SoSH9PxnSnyrXAyYAVknoW4G7u04vu50%2Bv6fhmX5wO1KLuZtDdVMfycNGhQ4zcVijYCPHY6ohjKsfNxdTPVS03VcqB3PR3DWWt0LDBgo0cXsL4dVSqmeaB1X5MUojgjoaza9jutFAptk%2BHGcJYaN6WGcYP%2FsP39KrNmhqwIgr8pNASQtBQrSLtTIItR89GUZXfODiHNUM%2FdZPad10A6PFnqVLiZ195hDByN%2FmyL4jIWfgFJE8wJhI1KHm0GY37sAYoxkP7NH3Bfr00QArKL0XB%2BL9i50GQa9lUIpgS8Iq8E7qgD4oJaT4Oe3gbMB9m4mS%2FmSGI%2Bj57o%2FO3Qct98oD%2FyzH9tJff65v8b3OyNnvTQLwLQbCXyCx1EGqhRD%2BHLFQCT%2B5kJGS5O9bKMzvGeMYpcnRVfjjmhBPEQ2v0Q%2F1rxHOhHC5SQapH2rUSVlAwo%2B467VwyanlqQnXkgnjtVkt%2Bw5roWvelWbuDrLypaap9gEXah9PfOS9KbRSCBDZxZKkgtitDqZHd7eCpaW3C20jqDs4rftNfDlZZA%2FlSIgLK46ERrTixrBvqBg2bTf484Zh6SvxG4bDKRdb9R%2FP1XsZP9pBKlLFxW3IjTy%2FeKybXcw3DmkcuM7JbwhAxVb2vD7YDjCSxsuPBjqlAVNDb7aL7uWi9Bz%2BNOKuubbQm%2FQ7ScxfWAqwn2thzH%2BGHkjByv%2BLUg4egwXVcWfK6j1LK4Wu%2FQMu8WLzQwllr2AN9QaOK3S9JXRCsdZMaqGGhMxOgRzd6q8WVNlrUAf8Xfti%2FgyjM3h4jU3Xclzd7ZBKFF8PhuQ6RCdKMQc%2BGW4H0SjWN9YqEPznqPDtchth9%2FAl7M20PDNSGvTUan4PNBDtUQOnag%3D%3D&Expires=1643440755 [following]\n",
      "--2022-01-27 11:19:15--  https://encode-public.s3.amazonaws.com/2021/02/24/57ce8c5b-7229-44b1-b5dc-e62d82a5b7d2/ENCFF987XOV.bam?response-content-disposition=attachment%3B%20filename%3DENCFF987XOV.bam&AWSAccessKeyId=ASIATGZNGCNXVI6AN3H7&Signature=UVv5jBum8NR4CIy1dN1y2ZUoEhc%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEJv%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJHMEUCIDim9%2FmXf7aVc2BrNnzI1trmTuAwp6%2F9kM5CwIRf0XDiAiEAjZG1lhLINFny2vUpvSVfsA9vbd6MuXY3pGgta5IgbnIqgwQIxP%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FARAAGgwyMjA3NDg3MTQ4NjMiDJMoTf9SoSH9PxnSnyrXAyYAVknoW4G7u04vu50%2Bv6fhmX5wO1KLuZtDdVMfycNGhQ4zcVijYCPHY6ohjKsfNxdTPVS03VcqB3PR3DWWt0LDBgo0cXsL4dVSqmeaB1X5MUojgjoaza9jutFAptk%2BHGcJYaN6WGcYP%2FsP39KrNmhqwIgr8pNASQtBQrSLtTIItR89GUZXfODiHNUM%2FdZPad10A6PFnqVLiZ195hDByN%2FmyL4jIWfgFJE8wJhI1KHm0GY37sAYoxkP7NH3Bfr00QArKL0XB%2BL9i50GQa9lUIpgS8Iq8E7qgD4oJaT4Oe3gbMB9m4mS%2FmSGI%2Bj57o%2FO3Qct98oD%2FyzH9tJff65v8b3OyNnvTQLwLQbCXyCx1EGqhRD%2BHLFQCT%2B5kJGS5O9bKMzvGeMYpcnRVfjjmhBPEQ2v0Q%2F1rxHOhHC5SQapH2rUSVlAwo%2B467VwyanlqQnXkgnjtVkt%2Bw5roWvelWbuDrLypaap9gEXah9PfOS9KbRSCBDZxZKkgtitDqZHd7eCpaW3C20jqDs4rftNfDlZZA%2FlSIgLK46ERrTixrBvqBg2bTf484Zh6SvxG4bDKRdb9R%2FP1XsZP9pBKlLFxW3IjTy%2FeKybXcw3DmkcuM7JbwhAxVb2vD7YDjCSxsuPBjqlAVNDb7aL7uWi9Bz%2BNOKuubbQm%2FQ7ScxfWAqwn2thzH%2BGHkjByv%2BLUg4egwXVcWfK6j1LK4Wu%2FQMu8WLzQwllr2AN9QaOK3S9JXRCsdZMaqGGhMxOgRzd6q8WVNlrUAf8Xfti%2FgyjM3h4jU3Xclzd7ZBKFF8PhuQ6RCdKMQc%2BGW4H0SjWN9YqEPznqPDtchth9%2FAl7M20PDNSGvTUan4PNBDtUQOnag%3D%3D&Expires=1643440755\n",
      "Resolving encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)... 52.218.153.235\n",
      "Connecting to encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)|52.218.153.235|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 2950424249 (2.7G) [binary/octet-stream]\n",
      "Saving to: ‘k562_rep2.bam’\n",
      "\n",
      "k562_rep2.bam       100%[===================>]   2.75G  25.4MB/s    in 2m 13s  \n",
      "\n",
      "2022-01-27 11:21:28 (21.2 MB/s) - ‘k562_rep2.bam’ saved [2950424249/2950424249]\n",
      "\n",
      "wget -nc -O hct116_rep1.bam \"https://www.encodeproject.org/files/ENCFF724QHH/@@download/ENCFF724QHH.bam\"\n",
      "--2022-01-27 11:21:28--  https://www.encodeproject.org/files/ENCFF724QHH/@@download/ENCFF724QHH.bam\n",
      "Resolving www.encodeproject.org (www.encodeproject.org)... 34.211.244.144\n",
      "Connecting to www.encodeproject.org (www.encodeproject.org)|34.211.244.144|:443... connected.\n",
      "HTTP request sent, awaiting response... 307 Temporary Redirect\n",
      "Location: https://encode-public.s3.amazonaws.com/2021/02/24/332385ae-9c78-4f08-8241-868f98f70848/ENCFF724QHH.bam?response-content-disposition=attachment%3B%20filename%3DENCFF724QHH.bam&AWSAccessKeyId=ASIATGZNGCNXVI6AN3H7&Signature=ZTCxPxrf7440yOzrOMLg5Viz1aI%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEJv%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJHMEUCIDim9%2FmXf7aVc2BrNnzI1trmTuAwp6%2F9kM5CwIRf0XDiAiEAjZG1lhLINFny2vUpvSVfsA9vbd6MuXY3pGgta5IgbnIqgwQIxP%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FARAAGgwyMjA3NDg3MTQ4NjMiDJMoTf9SoSH9PxnSnyrXAyYAVknoW4G7u04vu50%2Bv6fhmX5wO1KLuZtDdVMfycNGhQ4zcVijYCPHY6ohjKsfNxdTPVS03VcqB3PR3DWWt0LDBgo0cXsL4dVSqmeaB1X5MUojgjoaza9jutFAptk%2BHGcJYaN6WGcYP%2FsP39KrNmhqwIgr8pNASQtBQrSLtTIItR89GUZXfODiHNUM%2FdZPad10A6PFnqVLiZ195hDByN%2FmyL4jIWfgFJE8wJhI1KHm0GY37sAYoxkP7NH3Bfr00QArKL0XB%2BL9i50GQa9lUIpgS8Iq8E7qgD4oJaT4Oe3gbMB9m4mS%2FmSGI%2Bj57o%2FO3Qct98oD%2FyzH9tJff65v8b3OyNnvTQLwLQbCXyCx1EGqhRD%2BHLFQCT%2B5kJGS5O9bKMzvGeMYpcnRVfjjmhBPEQ2v0Q%2F1rxHOhHC5SQapH2rUSVlAwo%2B467VwyanlqQnXkgnjtVkt%2Bw5roWvelWbuDrLypaap9gEXah9PfOS9KbRSCBDZxZKkgtitDqZHd7eCpaW3C20jqDs4rftNfDlZZA%2FlSIgLK46ERrTixrBvqBg2bTf484Zh6SvxG4bDKRdb9R%2FP1XsZP9pBKlLFxW3IjTy%2FeKybXcw3DmkcuM7JbwhAxVb2vD7YDjCSxsuPBjqlAVNDb7aL7uWi9Bz%2BNOKuubbQm%2FQ7ScxfWAqwn2thzH%2BGHkjByv%2BLUg4egwXVcWfK6j1LK4Wu%2FQMu8WLzQwllr2AN9QaOK3S9JXRCsdZMaqGGhMxOgRzd6q8WVNlrUAf8Xfti%2FgyjM3h4jU3Xclzd7ZBKFF8PhuQ6RCdKMQc%2BGW4H0SjWN9YqEPznqPDtchth9%2FAl7M20PDNSGvTUan4PNBDtUQOnag%3D%3D&Expires=1643440888 [following]\n",
      "--2022-01-27 11:21:28--  https://encode-public.s3.amazonaws.com/2021/02/24/332385ae-9c78-4f08-8241-868f98f70848/ENCFF724QHH.bam?response-content-disposition=attachment%3B%20filename%3DENCFF724QHH.bam&AWSAccessKeyId=ASIATGZNGCNXVI6AN3H7&Signature=ZTCxPxrf7440yOzrOMLg5Viz1aI%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEJv%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJHMEUCIDim9%2FmXf7aVc2BrNnzI1trmTuAwp6%2F9kM5CwIRf0XDiAiEAjZG1lhLINFny2vUpvSVfsA9vbd6MuXY3pGgta5IgbnIqgwQIxP%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FARAAGgwyMjA3NDg3MTQ4NjMiDJMoTf9SoSH9PxnSnyrXAyYAVknoW4G7u04vu50%2Bv6fhmX5wO1KLuZtDdVMfycNGhQ4zcVijYCPHY6ohjKsfNxdTPVS03VcqB3PR3DWWt0LDBgo0cXsL4dVSqmeaB1X5MUojgjoaza9jutFAptk%2BHGcJYaN6WGcYP%2FsP39KrNmhqwIgr8pNASQtBQrSLtTIItR89GUZXfODiHNUM%2FdZPad10A6PFnqVLiZ195hDByN%2FmyL4jIWfgFJE8wJhI1KHm0GY37sAYoxkP7NH3Bfr00QArKL0XB%2BL9i50GQa9lUIpgS8Iq8E7qgD4oJaT4Oe3gbMB9m4mS%2FmSGI%2Bj57o%2FO3Qct98oD%2FyzH9tJff65v8b3OyNnvTQLwLQbCXyCx1EGqhRD%2BHLFQCT%2B5kJGS5O9bKMzvGeMYpcnRVfjjmhBPEQ2v0Q%2F1rxHOhHC5SQapH2rUSVlAwo%2B467VwyanlqQnXkgnjtVkt%2Bw5roWvelWbuDrLypaap9gEXah9PfOS9KbRSCBDZxZKkgtitDqZHd7eCpaW3C20jqDs4rftNfDlZZA%2FlSIgLK46ERrTixrBvqBg2bTf484Zh6SvxG4bDKRdb9R%2FP1XsZP9pBKlLFxW3IjTy%2FeKybXcw3DmkcuM7JbwhAxVb2vD7YDjCSxsuPBjqlAVNDb7aL7uWi9Bz%2BNOKuubbQm%2FQ7ScxfWAqwn2thzH%2BGHkjByv%2BLUg4egwXVcWfK6j1LK4Wu%2FQMu8WLzQwllr2AN9QaOK3S9JXRCsdZMaqGGhMxOgRzd6q8WVNlrUAf8Xfti%2FgyjM3h4jU3Xclzd7ZBKFF8PhuQ6RCdKMQc%2BGW4H0SjWN9YqEPznqPDtchth9%2FAl7M20PDNSGvTUan4PNBDtUQOnag%3D%3D&Expires=1643440888\n",
      "Resolving encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)... 52.218.219.35\n",
      "Connecting to encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)|52.218.219.35|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 3273232766 (3.0G) [binary/octet-stream]\n",
      "Saving to: ‘hct116_rep1.bam’\n",
      "\n",
      "hct116_rep1.bam     100%[===================>]   3.05G  32.3MB/s    in 2m 2s   \n",
      "\n",
      "2022-01-27 11:23:30 (25.7 MB/s) - ‘hct116_rep1.bam’ saved [3273232766/3273232766]\n",
      "\n",
      "wget -nc -O hct116_rep2.bam \"https://www.encodeproject.org/files/ENCFF927YUB/@@download/ENCFF927YUB.bam\"\n",
      "--2022-01-27 11:23:30--  https://www.encodeproject.org/files/ENCFF927YUB/@@download/ENCFF927YUB.bam\n",
      "Resolving www.encodeproject.org (www.encodeproject.org)... 34.211.244.144\n",
      "Connecting to www.encodeproject.org (www.encodeproject.org)|34.211.244.144|:443... connected.\n",
      "HTTP request sent, awaiting response... 307 Temporary Redirect\n",
      "Location: https://encode-public.s3.amazonaws.com/2021/02/24/ac76a7b8-09ec-48f7-98f8-18517b53578f/ENCFF927YUB.bam?response-content-disposition=attachment%3B%20filename%3DENCFF927YUB.bam&AWSAccessKeyId=ASIATGZNGCNXVI6AN3H7&Signature=nMs%2Bzdul0B1uFwNRAEEB2vvUnr0%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEJv%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJHMEUCIDim9%2FmXf7aVc2BrNnzI1trmTuAwp6%2F9kM5CwIRf0XDiAiEAjZG1lhLINFny2vUpvSVfsA9vbd6MuXY3pGgta5IgbnIqgwQIxP%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FARAAGgwyMjA3NDg3MTQ4NjMiDJMoTf9SoSH9PxnSnyrXAyYAVknoW4G7u04vu50%2Bv6fhmX5wO1KLuZtDdVMfycNGhQ4zcVijYCPHY6ohjKsfNxdTPVS03VcqB3PR3DWWt0LDBgo0cXsL4dVSqmeaB1X5MUojgjoaza9jutFAptk%2BHGcJYaN6WGcYP%2FsP39KrNmhqwIgr8pNASQtBQrSLtTIItR89GUZXfODiHNUM%2FdZPad10A6PFnqVLiZ195hDByN%2FmyL4jIWfgFJE8wJhI1KHm0GY37sAYoxkP7NH3Bfr00QArKL0XB%2BL9i50GQa9lUIpgS8Iq8E7qgD4oJaT4Oe3gbMB9m4mS%2FmSGI%2Bj57o%2FO3Qct98oD%2FyzH9tJff65v8b3OyNnvTQLwLQbCXyCx1EGqhRD%2BHLFQCT%2B5kJGS5O9bKMzvGeMYpcnRVfjjmhBPEQ2v0Q%2F1rxHOhHC5SQapH2rUSVlAwo%2B467VwyanlqQnXkgnjtVkt%2Bw5roWvelWbuDrLypaap9gEXah9PfOS9KbRSCBDZxZKkgtitDqZHd7eCpaW3C20jqDs4rftNfDlZZA%2FlSIgLK46ERrTixrBvqBg2bTf484Zh6SvxG4bDKRdb9R%2FP1XsZP9pBKlLFxW3IjTy%2FeKybXcw3DmkcuM7JbwhAxVb2vD7YDjCSxsuPBjqlAVNDb7aL7uWi9Bz%2BNOKuubbQm%2FQ7ScxfWAqwn2thzH%2BGHkjByv%2BLUg4egwXVcWfK6j1LK4Wu%2FQMu8WLzQwllr2AN9QaOK3S9JXRCsdZMaqGGhMxOgRzd6q8WVNlrUAf8Xfti%2FgyjM3h4jU3Xclzd7ZBKFF8PhuQ6RCdKMQc%2BGW4H0SjWN9YqEPznqPDtchth9%2FAl7M20PDNSGvTUan4PNBDtUQOnag%3D%3D&Expires=1643441010 [following]\n",
      "--2022-01-27 11:23:30--  https://encode-public.s3.amazonaws.com/2021/02/24/ac76a7b8-09ec-48f7-98f8-18517b53578f/ENCFF927YUB.bam?response-content-disposition=attachment%3B%20filename%3DENCFF927YUB.bam&AWSAccessKeyId=ASIATGZNGCNXVI6AN3H7&Signature=nMs%2Bzdul0B1uFwNRAEEB2vvUnr0%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEJv%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJHMEUCIDim9%2FmXf7aVc2BrNnzI1trmTuAwp6%2F9kM5CwIRf0XDiAiEAjZG1lhLINFny2vUpvSVfsA9vbd6MuXY3pGgta5IgbnIqgwQIxP%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FARAAGgwyMjA3NDg3MTQ4NjMiDJMoTf9SoSH9PxnSnyrXAyYAVknoW4G7u04vu50%2Bv6fhmX5wO1KLuZtDdVMfycNGhQ4zcVijYCPHY6ohjKsfNxdTPVS03VcqB3PR3DWWt0LDBgo0cXsL4dVSqmeaB1X5MUojgjoaza9jutFAptk%2BHGcJYaN6WGcYP%2FsP39KrNmhqwIgr8pNASQtBQrSLtTIItR89GUZXfODiHNUM%2FdZPad10A6PFnqVLiZ195hDByN%2FmyL4jIWfgFJE8wJhI1KHm0GY37sAYoxkP7NH3Bfr00QArKL0XB%2BL9i50GQa9lUIpgS8Iq8E7qgD4oJaT4Oe3gbMB9m4mS%2FmSGI%2Bj57o%2FO3Qct98oD%2FyzH9tJff65v8b3OyNnvTQLwLQbCXyCx1EGqhRD%2BHLFQCT%2B5kJGS5O9bKMzvGeMYpcnRVfjjmhBPEQ2v0Q%2F1rxHOhHC5SQapH2rUSVlAwo%2B467VwyanlqQnXkgnjtVkt%2Bw5roWvelWbuDrLypaap9gEXah9PfOS9KbRSCBDZxZKkgtitDqZHd7eCpaW3C20jqDs4rftNfDlZZA%2FlSIgLK46ERrTixrBvqBg2bTf484Zh6SvxG4bDKRdb9R%2FP1XsZP9pBKlLFxW3IjTy%2FeKybXcw3DmkcuM7JbwhAxVb2vD7YDjCSxsuPBjqlAVNDb7aL7uWi9Bz%2BNOKuubbQm%2FQ7ScxfWAqwn2thzH%2BGHkjByv%2BLUg4egwXVcWfK6j1LK4Wu%2FQMu8WLzQwllr2AN9QaOK3S9JXRCsdZMaqGGhMxOgRzd6q8WVNlrUAf8Xfti%2FgyjM3h4jU3Xclzd7ZBKFF8PhuQ6RCdKMQc%2BGW4H0SjWN9YqEPznqPDtchth9%2FAl7M20PDNSGvTUan4PNBDtUQOnag%3D%3D&Expires=1643441010\n",
      "Resolving encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)... 52.92.163.169\n",
      "Connecting to encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)|52.92.163.169|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 3471809963 (3.2G) [binary/octet-stream]\n",
      "Saving to: ‘hct116_rep2.bam’\n",
      "\n",
      "hct116_rep2.bam     100%[===================>]   3.23G  31.2MB/s    in 2m 12s  \n",
      "\n",
      "2022-01-27 11:25:42 (25.1 MB/s) - ‘hct116_rep2.bam’ saved [3471809963/3471809963]\n",
      "\n",
      "CPU times: user 6.82 s, sys: 1.65 s, total: 8.47 s\n",
      "Wall time: 10min 13s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "run_wget_bam_cmds = run_cmd = True\n",
    "for cmd in list(samplesheet_df['wget_bam_cmd']):\n",
    "    print(cmd)\n",
    "    if run_cmd:\n",
    "        ! {cmd}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abfadf77-6e55-40e8-a3ae-875aa0b3d934",
   "metadata": {},
   "source": [
    "# Index alignment files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "714c3208-92b0-4850-a6d3-8b997b9e9a3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "index_bam_threads = 8\n",
    "samplesheet_df['index_bam_cmd'] = (\n",
    "    'samtools index ' +\n",
    "    samplesheet_df['bam_filepath']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5dd876eb-4bf3-49eb-beb0-389e8a65ea5e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools index k562_rep1.bam\n",
      "samtools index k562_rep2.bam\n",
      "samtools index hct116_rep1.bam\n",
      "samtools index hct116_rep2.bam\n",
      "CPU times: user 885 ms, sys: 212 ms, total: 1.1 s\n",
      "Wall time: 1min 40s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "run_index_bam_cmds = run_cmd = True\n",
    "for cmd in list(samplesheet_df['index_bam_cmd']):\n",
    "    print(cmd)\n",
    "    if run_cmd:\n",
    "        ! {cmd}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea3505df-ba66-452b-93d0-f233a884420b",
   "metadata": {},
   "source": [
    "# Compute coverage bigwig files from alignment files\n",
    "You will need to use [effective genome size](https://deeptools.readthedocs.io/en/latest/content/feature/effectiveGenomeSize.html) numbers for the relevant genome.\n",
    "You could also use other normalizations of choice at this step. Here we use the deeptools `bamCoverage` defaults."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dce46351-8394-465d-a798-b7536581d3c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "effective_genome_size = 2913022398\n",
    "bamcoverage_binsize = 10\n",
    "bamcoverage_threads = 'max/2'\n",
    "# Value for GRCh38, from:\n",
    "# https://deeptools.readthedocs.io/en/latest/content/feature/effectiveGenomeSize.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e2dc7c21-5bc0-4d8e-ae2e-9d05d82800aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "samplesheet_df['bw_filepath'] = samplesheet_df['basename'] + '.bw'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "42725da3-612d-4647-b43d-32e229bfea4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "samplesheet_df['bamcoverage_cmd'] = (\n",
    "    f'bamCoverage ' + \n",
    "    f' -p {bamcoverage_threads} ' + \n",
    "    f' --effectiveGenomeSize {effective_genome_size} ' + \n",
    "    f' -bs {bamcoverage_binsize}'\n",
    "    f' -b ' + samplesheet_df['bam_filepath'] + \n",
    "    f' -o ' + samplesheet_df['bw_filepath']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "23451f55-cd66-41b9-93d7-56c9e95320a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bamCoverage  -p max/2  --effectiveGenomeSize 2913022398  -bs 10 -b k562_rep1.bam -o k562_rep1.bw\n",
      "bamFilesList: ['k562_rep1.bam']\n",
      "binLength: 10\n",
      "numberOfSamples: None\n",
      "blackListFileName: None\n",
      "skipZeroOverZero: False\n",
      "bed_and_bin: False\n",
      "genomeChunkSize: None\n",
      "defaultFragmentLength: read length\n",
      "numberOfProcessors: 28\n",
      "verbose: False\n",
      "region: None\n",
      "bedFile: None\n",
      "minMappingQuality: None\n",
      "ignoreDuplicates: False\n",
      "chrsToSkip: []\n",
      "stepSize: 10\n",
      "center_read: False\n",
      "samFlag_include: None\n",
      "samFlag_exclude: None\n",
      "minFragmentLength: 0\n",
      "maxFragmentLength: 0\n",
      "zerosToNans: False\n",
      "smoothLength: None\n",
      "save_data: False\n",
      "out_file_for_raw_data: None\n",
      "maxPairedFragmentLength: 1000\n",
      "bamCoverage  -p max/2  --effectiveGenomeSize 2913022398  -bs 10 -b k562_rep2.bam -o k562_rep2.bw\n",
      "bamFilesList: ['k562_rep2.bam']\n",
      "binLength: 10\n",
      "numberOfSamples: None\n",
      "blackListFileName: None\n",
      "skipZeroOverZero: False\n",
      "bed_and_bin: False\n",
      "genomeChunkSize: None\n",
      "defaultFragmentLength: read length\n",
      "numberOfProcessors: 28\n",
      "verbose: False\n",
      "region: None\n",
      "bedFile: None\n",
      "minMappingQuality: None\n",
      "ignoreDuplicates: False\n",
      "chrsToSkip: []\n",
      "stepSize: 10\n",
      "center_read: False\n",
      "samFlag_include: None\n",
      "samFlag_exclude: None\n",
      "minFragmentLength: 0\n",
      "maxFragmentLength: 0\n",
      "zerosToNans: False\n",
      "smoothLength: None\n",
      "save_data: False\n",
      "out_file_for_raw_data: None\n",
      "maxPairedFragmentLength: 1000\n",
      "bamCoverage  -p max/2  --effectiveGenomeSize 2913022398  -bs 10 -b hct116_rep1.bam -o hct116_rep1.bw\n",
      "bamFilesList: ['hct116_rep1.bam']\n",
      "binLength: 10\n",
      "numberOfSamples: None\n",
      "blackListFileName: None\n",
      "skipZeroOverZero: False\n",
      "bed_and_bin: False\n",
      "genomeChunkSize: None\n",
      "defaultFragmentLength: read length\n",
      "numberOfProcessors: 28\n",
      "verbose: False\n",
      "region: None\n",
      "bedFile: None\n",
      "minMappingQuality: None\n",
      "ignoreDuplicates: False\n",
      "chrsToSkip: []\n",
      "stepSize: 10\n",
      "center_read: False\n",
      "samFlag_include: None\n",
      "samFlag_exclude: None\n",
      "minFragmentLength: 0\n",
      "maxFragmentLength: 0\n",
      "zerosToNans: False\n",
      "smoothLength: None\n",
      "save_data: False\n",
      "out_file_for_raw_data: None\n",
      "maxPairedFragmentLength: 1000\n",
      "bamCoverage  -p max/2  --effectiveGenomeSize 2913022398  -bs 10 -b hct116_rep2.bam -o hct116_rep2.bw\n",
      "bamFilesList: ['hct116_rep2.bam']\n",
      "binLength: 10\n",
      "numberOfSamples: None\n",
      "blackListFileName: None\n",
      "skipZeroOverZero: False\n",
      "bed_and_bin: False\n",
      "genomeChunkSize: None\n",
      "defaultFragmentLength: read length\n",
      "numberOfProcessors: 28\n",
      "verbose: False\n",
      "region: None\n",
      "bedFile: None\n",
      "minMappingQuality: None\n",
      "ignoreDuplicates: False\n",
      "chrsToSkip: []\n",
      "stepSize: 10\n",
      "center_read: False\n",
      "samFlag_include: None\n",
      "samFlag_exclude: None\n",
      "minFragmentLength: 0\n",
      "maxFragmentLength: 0\n",
      "zerosToNans: False\n",
      "smoothLength: None\n",
      "save_data: False\n",
      "out_file_for_raw_data: None\n",
      "maxPairedFragmentLength: 1000\n",
      "CPU times: user 5.54 s, sys: 1.19 s, total: 6.73 s\n",
      "Wall time: 9min 27s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "run_bamcoverage_cmds = run_cmd = True\n",
    "for cmd in list(samplesheet_df['bamcoverage_cmd']):\n",
    "    print(cmd)\n",
    "    if run_cmd:\n",
    "        ! {cmd}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b25f5994-c223-4d61-b18d-8209e8f1e3ac",
   "metadata": {},
   "source": [
    "# Download reference genome\n",
    "Also generate index and chromosome size files for bedtools and wigToBigWig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a7d2ffb4-53dc-481b-9036-0c61b9629f4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_fa_filepath = 'hg38.fa'\n",
    "genome_fai_filepath = f'{genome_fa_filepath}.fai'\n",
    "genome_chromsizes_filepath = f'{genome_fa_filepath}.chromsizes.tab'\n",
    "genome_fa_gz_url = 'https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/latest/hg38.fa.masked.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "13410ecd-5042-4aeb-becb-731a84630dae",
   "metadata": {},
   "outputs": [],
   "source": [
    "download_genome_fa_cmd = f'wget -nc -O {genome_fa_filepath}.gz \"{genome_fa_gz_url}\"; zcat {genome_fa_filepath}.gz > {genome_fa_filepath}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b03a18ea-b60d-4080-b3d8-5859c24c2124",
   "metadata": {},
   "outputs": [],
   "source": [
    "index_genome_fa_cmd = f'samtools faidx {genome_fa_filepath}'\n",
    "genome_chromsizes_cmd = f'cut -f1,2 {genome_fai_filepath} > {genome_chromsizes_filepath}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "39a2ddff-6dcd-46d9-9854-6299b7ff89c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wget -nc -O hg38.fa.gz \"https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/latest/hg38.fa.masked.gz\"; zcat hg38.fa.gz > hg38.fa\n",
      "--2022-01-27 11:36:50--  https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/latest/hg38.fa.masked.gz\n",
      "Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.119.163\n",
      "Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.119.163|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 510854439 (487M) [application/x-gzip]\n",
      "Saving to: ‘hg38.fa.gz’\n",
      "\n",
      "hg38.fa.gz          100%[===================>] 487.19M  86.2MB/s    in 6.4s    \n",
      "\n",
      "2022-01-27 11:36:56 (75.7 MB/s) - ‘hg38.fa.gz’ saved [510854439/510854439]\n",
      "\n",
      "samtools faidx hg38.fa\n",
      "cut -f1,2 hg38.fa.fai > hg38.fa.chromsizes.tab\n",
      "CPU times: user 323 ms, sys: 103 ms, total: 426 ms\n",
      "Wall time: 36.8 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "run_download_genome_fa_cmd = run_cmd = True\n",
    "for cmd in [download_genome_fa_cmd, index_genome_fa_cmd, genome_chromsizes_cmd]:\n",
    "    print(cmd)\n",
    "    if run_cmd:\n",
    "        ! {cmd}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e77dee5-fbde-45c7-8049-07e634fb85c5",
   "metadata": {},
   "source": [
    "# Designate comparison\n",
    "Here we compare HCT116 vs. K562 cells. Comparison groups are designated by the `cell_type` column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ef218e4d-482f-4ba3-9c9b-241d5249756f",
   "metadata": {},
   "outputs": [],
   "source": [
    "group_1  = 'hct116'\n",
    "group_2  = 'k562'\n",
    "\n",
    "group_column = 'cell_type'\n",
    "sample_column = 'sample'\n",
    "sort_column = 'replicate'\n",
    "\n",
    "comparison_prefix = f'{group_1}.vs.{group_2}'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "227de0fa-c04d-43cb-b700-a6ec5a0635e6",
   "metadata": {},
   "source": [
    "# List bigwigs belonging to each group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "51771554-b11a-44bd-be0f-047189e8d4d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_subset_df = samplesheet_df[samplesheet_df[sample_column]].sort_values(by = sort_column).copy()\n",
    "\n",
    "group_1_sample_bw_filepaths = list(sample_subset_df[sample_subset_df[group_column] == group_1]['bw_filepath'])\n",
    "group_2_sample_bw_filepaths = list(sample_subset_df[sample_subset_df[group_column] == group_2]['bw_filepath'])\n",
    "sample_bw_filepaths = group_1_sample_bw_filepaths + group_2_sample_bw_filepaths\n",
    "group_1_sample_bw_filepaths_str = ' '.join(group_1_sample_bw_filepaths)\n",
    "group_2_sample_bw_filepaths_str = ' '.join(group_2_sample_bw_filepaths)\n",
    "sample_bw_filepaths_str = ' '.join(sample_bw_filepaths)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d389772e-2516-46ee-bd15-810f7fe113bb",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Calculate Bigwig of Log2FC between groups.\n",
    "First we calculate the means of each group, in `{group_1}.mean.bw` and `{group_2}.mean.bw`. Then we compute the Log2 Fold Change (with pseudocount) as `log2((group_1_mean+1)/(group_2_mean+1))`. A pseudocount prevents division by zero in the ratio calculation.\n",
    "\n",
    "We also compute the sum of coverage across all samples, for use later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "07be3510-5e77-4fbb-960a-78fbb10188a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "group_1_bw_filepath = f'{group_1}.mean.bw'\n",
    "group_2_bw_filepath = f'{group_2}.mean.bw'\n",
    "\n",
    "sum_bw_filepath = f'{group_1}.vs.{group_2}.sum.bw'\n",
    "log2fc_bw_filepath = f'{group_1}.vs.{group_2}.log2fc.bw'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "23a7ae13-4684-4ef7-a187-599d0a7b0bde",
   "metadata": {},
   "outputs": [],
   "source": [
    "group_1_bw_cmd = (\n",
    "    f'wiggletools write {group_1_bw_filepath}.wig mean {group_1_sample_bw_filepaths_str} ; '\n",
    "    f'wigToBigWig -clip {group_1_bw_filepath}.wig {genome_chromsizes_filepath} {group_1_bw_filepath} ; '\n",
    "    f'rm {group_1_bw_filepath}.wig'\n",
    ")\n",
    "group_2_bw_cmd = (\n",
    "    f'wiggletools write {group_2_bw_filepath}.wig mean {group_2_sample_bw_filepaths_str} ; '\n",
    "    f'wigToBigWig -clip {group_2_bw_filepath}.wig {genome_chromsizes_filepath} {group_2_bw_filepath} ; '\n",
    "    f'rm {group_2_bw_filepath}.wig'\n",
    ")\n",
    "\n",
    "sum_bw_cmd = (\n",
    "    f'wiggletools write {sum_bw_filepath}.wig sum {group_1_sample_bw_filepaths_str} {group_2_sample_bw_filepaths_str} ; '\n",
    "    f'wigToBigWig -clip {sum_bw_filepath}.wig {genome_chromsizes_filepath} {sum_bw_filepath} ; '\n",
    "    f'rm {sum_bw_filepath}.wig'\n",
    ")\n",
    "\n",
    "log2fc_bw_cmd = (\n",
    "    f'wiggletools write {group_1_bw_filepath}.plus_1.wig offset 1 {group_1_bw_filepath} ; '\n",
    "    f'wiggletools write {group_2_bw_filepath}.plus_1.wig offset 1 {group_2_bw_filepath} ; '\n",
    "    f'wiggletools write {group_1}.vs.{group_2}.ratio.wig ratio {group_1_bw_filepath}.plus_1.wig {group_2_bw_filepath}.plus_1.wig ; '\n",
    "    f'wiggletools write {log2fc_bw_filepath}.wig log 2 {group_1}.vs.{group_2}.ratio.wig ; '\n",
    "    f'wigToBigWig -clip {log2fc_bw_filepath}.wig {genome_chromsizes_filepath} {log2fc_bw_filepath} ; '\n",
    "    f'rm {group_1_bw_filepath}.plus_1.wig ; '\n",
    "    f'rm {group_2_bw_filepath}.plus_1.wig ; '\n",
    "    f'rm {group_1}.vs.{group_2}.ratio.wig ; '\n",
    "    f'rm {log2fc_bw_filepath}.wig '\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "0f277ed1-62f3-43b9-9be5-35720ce40e5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "bw_cmds = [group_1_bw_cmd, group_2_bw_cmd, sum_bw_cmd, log2fc_bw_cmd]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "324a9c4c-570a-4e42-b3da-2d5c4fd71c85",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wiggletools write hct116.mean.bw.wig mean hct116_rep1.bw hct116_rep2.bw ; wigToBigWig -clip hct116.mean.bw.wig hg38.fa.chromsizes.tab hct116.mean.bw ; rm hct116.mean.bw.wig\n",
      "chromosome chrEBV is not in chrom sizes file\n",
      "chromosome chrEBV is not in chrom sizes file\n",
      "wiggletools write k562.mean.bw.wig mean k562_rep1.bw k562_rep2.bw ; wigToBigWig -clip k562.mean.bw.wig hg38.fa.chromsizes.tab k562.mean.bw ; rm k562.mean.bw.wig\n",
      "chromosome chrEBV is not in chrom sizes file\n",
      "chromosome chrEBV is not in chrom sizes file\n",
      "wiggletools write hct116.vs.k562.sum.bw.wig sum hct116_rep1.bw hct116_rep2.bw k562_rep1.bw k562_rep2.bw ; wigToBigWig -clip hct116.vs.k562.sum.bw.wig hg38.fa.chromsizes.tab hct116.vs.k562.sum.bw ; rm hct116.vs.k562.sum.bw.wig\n",
      "chromosome chrEBV is not in chrom sizes file\n",
      "chromosome chrEBV is not in chrom sizes file\n",
      "wiggletools write hct116.mean.bw.plus_1.wig offset 1 hct116.mean.bw ; wiggletools write k562.mean.bw.plus_1.wig offset 1 k562.mean.bw ; wiggletools write hct116.vs.k562.ratio.wig ratio hct116.mean.bw.plus_1.wig k562.mean.bw.plus_1.wig ; wiggletools write hct116.vs.k562.log2fc.bw.wig log 2 hct116.vs.k562.ratio.wig ; wigToBigWig -clip hct116.vs.k562.log2fc.bw.wig hg38.fa.chromsizes.tab hct116.vs.k562.log2fc.bw ; rm hct116.mean.bw.plus_1.wig ; rm k562.mean.bw.plus_1.wig ; rm hct116.vs.k562.ratio.wig ; rm hct116.vs.k562.log2fc.bw.wig \n",
      "chromosome chrEBV is not in chrom sizes file\n",
      "chromosome chrEBV is not in chrom sizes file\n",
      "CPU times: user 7.73 s, sys: 1.76 s, total: 9.49 s\n",
      "Wall time: 14min 17s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "run_bw_cmds = run_cmd = True\n",
    "for cmd in list(bw_cmds):\n",
    "    print(cmd)\n",
    "    if run_cmd:\n",
    "        ! {cmd}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4efd989f-b120-4e6d-b7f6-98841ba25f9e",
   "metadata": {},
   "source": [
    "# Download annotation file\n",
    "Not strictly necessary, just for the benefit of visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c9fbbce4-a78a-4d4d-9e78-2332bbc9d225",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_gtf_filepath = 'hg38.ensGene.gtf'\n",
    "genes_gtf_gz_url = 'https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/hg38.ensGene.gtf.gz'\n",
    "\n",
    "# genes_gtf_filepath = 'ENCFF159KBI.gtf'\n",
    "# genes_gtf_gz_url = 'https://www.encodeproject.org/files/ENCFF159KBI/@@download/ENCFF159KBI.gtf.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "7962b7c9-c96d-4f5a-acaf-40f109aa5203",
   "metadata": {},
   "outputs": [],
   "source": [
    "download_genes_gtf_cmd = f'wget -nc -O {genes_gtf_filepath}.gz \"{genes_gtf_gz_url}\"; zcat {genes_gtf_filepath}.gz > {genes_gtf_filepath}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d2b57e72-0918-438d-89c9-687bc08891e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wget -nc -O hg38.ensGene.gtf.gz \"https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/hg38.ensGene.gtf.gz\"; zcat hg38.ensGene.gtf.gz > hg38.ensGene.gtf\n",
      "--2022-01-27 11:51:44--  https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/hg38.ensGene.gtf.gz\n",
      "Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.119.163\n",
      "Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.119.163|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 27802050 (27M) [application/x-gzip]\n",
      "Saving to: ‘hg38.ensGene.gtf.gz’\n",
      "\n",
      "hg38.ensGene.gtf.gz 100%[===================>]  26.51M  38.3MB/s    in 0.7s    \n",
      "\n",
      "2022-01-27 11:51:45 (38.3 MB/s) - ‘hg38.ensGene.gtf.gz’ saved [27802050/27802050]\n",
      "\n",
      "CPU times: user 35.6 ms, sys: 7.68 ms, total: 43.3 ms\n",
      "Wall time: 2.72 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "run_download_genes_gtf_cmd = run_cmd = True\n",
    "for cmd in [download_genes_gtf_cmd]:\n",
    "    print(cmd)\n",
    "    if run_cmd:\n",
    "        ! {cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "88b1de51-1c57-4622-9a1e-6887aea224f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\tensGene\ttranscript\t11869\t14409\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000456328\";  gene_name \"ENSG00000223972\";\n",
      "chr1\tensGene\texon\t11869\t12227\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000456328\"; exon_number \"1\"; exon_id \"ENST00000456328.1\"; gene_name \"ENSG00000223972\";\n",
      "chr1\tensGene\texon\t12613\t12721\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000456328\"; exon_number \"2\"; exon_id \"ENST00000456328.2\"; gene_name \"ENSG00000223972\";\n",
      "chr1\tensGene\texon\t13221\t14409\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000456328\"; exon_number \"3\"; exon_id \"ENST00000456328.3\"; gene_name \"ENSG00000223972\";\n",
      "chr1\tensGene\ttranscript\t12010\t13670\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000450305\";  gene_name \"ENSG00000223972\";\n",
      "chr1\tensGene\texon\t12010\t12057\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000450305\"; exon_number \"1\"; exon_id \"ENST00000450305.1\"; gene_name \"ENSG00000223972\";\n",
      "chr1\tensGene\texon\t12179\t12227\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000450305\"; exon_number \"2\"; exon_id \"ENST00000450305.2\"; gene_name \"ENSG00000223972\";\n",
      "chr1\tensGene\texon\t12613\t12697\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000450305\"; exon_number \"3\"; exon_id \"ENST00000450305.3\"; gene_name \"ENSG00000223972\";\n",
      "chr1\tensGene\texon\t12975\t13052\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000450305\"; exon_number \"4\"; exon_id \"ENST00000450305.4\"; gene_name \"ENSG00000223972\";\n",
      "chr1\tensGene\texon\t13221\t13374\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000450305\"; exon_number \"5\"; exon_id \"ENST00000450305.5\"; gene_name \"ENSG00000223972\";\n"
     ]
    }
   ],
   "source": [
    "! head {genes_gtf_filepath}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6edb1e4-1bea-4fde-ac36-7e696187ccfe",
   "metadata": {},
   "source": [
    "# Visualize Log2 Fold Change bigWig with Coolbox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "af7ffae9-2f36-4f95-824c-2a3ea4e6812e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import coolbox\n",
    "from coolbox.api import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "0d4326f2-8f3a-4859-9697-06c991c38de3",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_range = 'chr3:135584655-135691546'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f3b5a9d1-383f-45be-89da-5631e7619b3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7acde7ca70a940edaa208ca28ad0bbb7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(VBox(children=(HBox(children=(Dropdown(options=('chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6',…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "frame = (\n",
    "    XAxis() + \n",
    "    GTF(genes_gtf_filepath, row_filter=\"feature == 'gene'\", name_attr='gene_name') + Title('Genes') + TrackHeight(5) + Color('#000000') +\n",
    "    BigWig(log2fc_bw_filepath) + Title('Log2FC') + Color('#cf32cf') +\n",
    "    BigWig(group_1_bw_filepath) + Title(group_1) + Color('#3232cf') +\n",
    "    BigWig(group_2_bw_filepath) + Title(group_2) + Color('#cf3232')\n",
    ")\n",
    "bsr = Browser(frame)\n",
    "frame.goto(test_range)\n",
    "bsr.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fc7bec0-d5a3-488b-82c4-af9be15ac54c",
   "metadata": {},
   "source": [
    "# Copy over the HOMER motif library\n",
    "Use the below commands to locate and copy your motif library if you installed HOMER with conda/mamba"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1b376dc3-d317-429b-8815-6d171d91416f",
   "metadata": {},
   "outputs": [],
   "source": [
    "copy_homer_motif_cmd = f'cp -rf $(dirname $(which homer))/../share/homer/motifs ./homer_motifs'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c59072b4-d87b-4cc1-9331-461070f34851",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cp -rf $(dirname $(which homer))/../share/homer/motifs ./homer_motifs\n",
      "CPU times: user 13.5 ms, sys: 11.2 ms, total: 24.8 ms\n",
      "Wall time: 1.34 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "run_copy_homer_motif_cmd = run_cmd = True\n",
    "for cmd in [copy_homer_motif_cmd]:\n",
    "    print(cmd)\n",
    "    if run_cmd:\n",
    "        ! {cmd}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "433bb49b-4d05-4a52-80b4-ae178aebce43",
   "metadata": {},
   "source": [
    "# Scan for motif instances to center sequence on"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "f4e7ba3d-28d2-4411-b570-29f4ce0626d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_basename = 'gata'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "bfe9ce2a-5bd1-4d08-a7d2-709efbdaae85",
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_subpath = '/'.join(motif_basename.split(' '))\n",
    "motif_safe_basename = '_'.join(motif_subpath.split('/'))\n",
    "scanned_motif_filepath = f'homer_motifs/{motif_subpath}.motif'\n",
    "motif_scans_bed_filepath = f'{motif_safe_basename}.scans.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6d68860b-edad-46fa-93fa-37874a9ae250",
   "metadata": {},
   "outputs": [],
   "source": [
    "scan_motifs_cmd = (\n",
    "    f'scanMotifGenomeWide.pl {scanned_motif_filepath} {genome_fa_filepath} '\n",
    "    f'-bed -5p 1> {motif_scans_bed_filepath} 2> {motif_scans_bed_filepath}.log'\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "29181289-d557-4574-ad81-a584670db863",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanMotifGenomeWide.pl homer_motifs/gata.motif hg38.fa -bed -5p 1> gata.scans.bed 2> gata.scans.bed.log\n",
      "CPU times: user 1.93 s, sys: 400 ms, total: 2.33 s\n",
      "Wall time: 3min 54s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "run_scan_motifs_cmd = run_cmd = True\n",
    "for cmd in [scan_motifs_cmd]:\n",
    "    print(cmd)\n",
    "    if run_cmd:\n",
    "        ! {cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "f58d0be4-a9b6-4777-b51f-e504d1dd9feb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">CAGATAAGGN\tGata1(Zf)/K562-GATA1-ChIP-Seq(GSE18829)/Homer\t6.402981\t-6.924523e+03\t0\t50000.0,9843.0,7580.0,5717.0,0.00e+00\n",
      "0.109\t0.469\t0.365\t0.057\n",
      "0.700\t0.024\t0.001\t0.275\n",
      "0.001\t0.001\t0.997\t0.001\n",
      "0.997\t0.001\t0.001\t0.001\n",
      "0.001\t0.001\t0.001\t0.997\n",
      "0.997\t0.001\t0.001\t0.001\n",
      "0.997\t0.001\t0.001\t0.001\n",
      "0.023\t0.145\t0.830\t0.001\n",
      "0.311\t0.237\t0.434\t0.018\n"
     ]
    }
   ],
   "source": [
    "! head {scanned_motif_filepath}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "fbba6a01-c07c-4d18-957c-a4c398e30391",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t22919\t23119\tGata1(Zf)\t8.474400\t+\n",
      "chr1\t27984\t28184\tGata1(Zf)\t8.166325\t-\n",
      "chr1\t32492\t32692\tGata1(Zf)\t8.746133\t+\n",
      "chr1\t32657\t32857\tGata1(Zf)\t7.109025\t-\n",
      "chr1\t42233\t42433\tGata1(Zf)\t8.746133\t+\n",
      "chr1\t48006\t48206\tGata1(Zf)\t8.153403\t-\n",
      "chr1\t50012\t50212\tGata1(Zf)\t9.838112\t-\n",
      "chr1\t51059\t51259\tGata1(Zf)\t6.645415\t-\n",
      "chr1\t53345\t53545\tGata1(Zf)\t9.020373\t+\n",
      "chr1\t53472\t53672\tGata1(Zf)\t9.954682\t+\n",
      "525556 gata.scans.bed\n"
     ]
    }
   ],
   "source": [
    "! head {motif_scans_bed_filepath}\n",
    "! wc -l {motif_scans_bed_filepath}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "638bf791-45eb-4e2f-9e15-7f8c74aea523",
   "metadata": {},
   "source": [
    "# Score motif scans by Log2 Fold Change, and annotate with coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "de8b4bf1-2506-4ea7-9bdf-6595f3451d6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "scored_motif_scans_filepath = f'{motif_safe_basename}.scans.scored_by.{log2fc_bw_filepath}.bed'\n",
    "coverage_summed_motif_scans_filepath = f'{motif_safe_basename}.scans.scored_by.{sum_bw_filepath}.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "42216ba9-b34c-4de7-8073-7bcd534fa3f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "score_scans_cmd = f'bigWigToWig {log2fc_bw_filepath} >(wig2bed -x) | bedmap --echo --delim \\'\\\\t\\' --wmean {motif_scans_bed_filepath} - | awk \\'$7!=\"NAN\"\\' | awk \\'{{FS=OFS=\"\\\\t\";$5=$7;print $1,$2,$3,$4,$5,$6}}\\' > {scored_motif_scans_filepath}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "aacb5cfb-c2c3-451b-bda2-11366bdbb4fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "coverage_sum_scans_cmd = f'bigWigToWig {sum_bw_filepath} >(wig2bed -x) | bedmap --echo --delim \\'\\\\t\\' --wmean {scored_motif_scans_filepath} - > {coverage_summed_motif_scans_filepath}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "a2a8ceb4-f60c-4011-ae66-65bfcc65becb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bigWigToWig hct116.vs.k562.log2fc.bw >(wig2bed -x) | bedmap --echo --delim '\\t' --wmean gata.scans.bed - | awk '$7!=\"NAN\"' | awk '{FS=OFS=\"\\t\";$5=$7;print $1,$2,$3,$4,$5,$6}' > gata.scans.scored_by.hct116.vs.k562.log2fc.bw.bed\n",
      "bigWigToWig hct116.vs.k562.sum.bw >(wig2bed -x) | bedmap --echo --delim '\\t' --wmean gata.scans.scored_by.hct116.vs.k562.log2fc.bw.bed - > gata.scans.scored_by.hct116.vs.k562.sum.bw.bed\n",
      "CPU times: user 4.48 s, sys: 934 ms, total: 5.41 s\n",
      "Wall time: 8min 28s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "run_score_scans_cmds = run_cmd = True\n",
    "for cmd in [score_scans_cmd, coverage_sum_scans_cmd]:\n",
    "    print(cmd)\n",
    "    if run_cmd:\n",
    "        ! {cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "35083c82-55c6-4dcc-afcb-2d3573d5ff28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t22919\t23119\tGata1(Zf)\t0.000000\t+\n",
      "chr1\t27984\t28184\tGata1(Zf)\t0.000000\t-\n",
      "chr1\t32492\t32692\tGata1(Zf)\t0.000000\t+\n",
      "chr1\t32657\t32857\tGata1(Zf)\t-0.839098\t-\n",
      "chr1\t42233\t42433\tGata1(Zf)\t-0.983107\t+\n",
      "chr1\t48006\t48206\tGata1(Zf)\t-0.403047\t-\n",
      "chr1\t50012\t50212\tGata1(Zf)\t1.275901\t-\n",
      "chr1\t51059\t51259\tGata1(Zf)\t-0.230130\t-\n",
      "chr1\t53345\t53545\tGata1(Zf)\t-1.404719\t+\n",
      "chr1\t53472\t53672\tGata1(Zf)\t-1.393854\t+\n"
     ]
    }
   ],
   "source": [
    "! head {scored_motif_scans_filepath}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "30c89f9b-e0c1-440b-adc3-25302b009d86",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t22919\t23119\tGata1(Zf)\t0.000000\t+\t0.000000\n",
      "chr1\t27984\t28184\tGata1(Zf)\t0.000000\t-\t0.000000\n",
      "chr1\t32492\t32692\tGata1(Zf)\t0.000000\t+\t0.000000\n",
      "chr1\t32657\t32857\tGata1(Zf)\t-0.839098\t-\t2.175000\n",
      "chr1\t42233\t42433\tGata1(Zf)\t-0.983107\t+\t18.085000\n",
      "chr1\t48006\t48206\tGata1(Zf)\t-0.403047\t-\t2.710000\n",
      "chr1\t50012\t50212\tGata1(Zf)\t1.275901\t-\t2.940000\n",
      "chr1\t51059\t51259\tGata1(Zf)\t-0.230130\t-\t6.040000\n",
      "chr1\t53345\t53545\tGata1(Zf)\t-1.404719\t+\t8.875000\n",
      "chr1\t53472\t53672\tGata1(Zf)\t-1.393854\t+\t5.430000\n"
     ]
    }
   ],
   "source": [
    "! head {coverage_summed_motif_scans_filepath}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac5c7478-f4be-4d82-9e9b-c2f53759f1ef",
   "metadata": {},
   "source": [
    "# Perform cluster deduplication, designate sequence length\n",
    "By default, HOMER genomewide motif scans extract intervals +/-100bp of the motif. This step prevents extraction of overlapping sequences in this interval, which prevents e.g. identification of artifactual periodicities/positionalities due to repetitive sequence.\n",
    "\n",
    "Briefly, we cluster overlapping intervals, then for each cluster we select only the interval with the highest summed coverage across all samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "3538be3b-2f96-4a85-915b-c727b7614adf",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence_length = 200"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "a5fd89c0-1c21-463e-b148-0743270294e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence_length = max(sequence_length, 200)\n",
    "slop = (sequence_length-200)//2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "400ffa92-5f26-4e7d-9761-5c0e8e82d194",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_deduplicated_scored_motif_scans_filepath = scored_motif_scans_filepath[:-len('.bed')] + '.cluster_deduplicated.bed'\n",
    "slopped_cluster_deduplicated_scored_motif_scans_filepath = cluster_deduplicated_scored_motif_scans_filepath[:-len('.bed')] + f'.slop_{slop}.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "4a5aa1c3-95c6-4851-94b1-17687ab3961b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_deduplication_cmd = (\n",
    "    f'bedtools cluster -s -i {coverage_summed_motif_scans_filepath} '\n",
    "    f'| sort -k8,8n -k7,7nr | awk \\'!a[$8]++\\' '\n",
    "    f'| bedtools sort -i - |cut -f1-6 '\n",
    "    f'> {cluster_deduplicated_scored_motif_scans_filepath}'\n",
    ")\n",
    "\n",
    "slop_cmd = (\n",
    "    f'bedtools slop -i {cluster_deduplicated_scored_motif_scans_filepath} -b {slop} -g {genome_chromsizes_filepath} > {slopped_cluster_deduplicated_scored_motif_scans_filepath}'\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "1c7477dd-f3a0-4940-b046-63b4a59477ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bedtools cluster -s -i gata.scans.scored_by.hct116.vs.k562.sum.bw.bed | sort -k8,8n -k7,7nr | awk '!a[$8]++' | bedtools sort -i - |cut -f1-6 > gata.scans.scored_by.hct116.vs.k562.log2fc.bw.cluster_deduplicated.bed\n",
      "bedtools slop -i gata.scans.scored_by.hct116.vs.k562.log2fc.bw.cluster_deduplicated.bed -b 0 -g hg38.fa.chromsizes.tab > gata.scans.scored_by.hct116.vs.k562.log2fc.bw.cluster_deduplicated.slop_0.bed\n",
      "CPU times: user 53.6 ms, sys: 32.1 ms, total: 85.7 ms\n",
      "Wall time: 6.49 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "run_bedtools_cmds = run_cmd = True\n",
    "for cmd in [cluster_deduplication_cmd, slop_cmd]:\n",
    "    print(cmd)\n",
    "    if run_cmd:\n",
    "        ! {cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "c7ae0fc4-8b09-486d-8fe8-047d2d6c75db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t22919\t23119\tGata1(Zf)\t0.000000\t+\n",
      "chr1\t27984\t28184\tGata1(Zf)\t0.000000\t-\n",
      "chr1\t32492\t32692\tGata1(Zf)\t0.000000\t+\n",
      "chr1\t32657\t32857\tGata1(Zf)\t-0.839098\t-\n",
      "chr1\t42233\t42433\tGata1(Zf)\t-0.983107\t+\n",
      "chr1\t48006\t48206\tGata1(Zf)\t-0.403047\t-\n",
      "chr1\t50012\t50212\tGata1(Zf)\t1.275901\t-\n",
      "chr1\t51059\t51259\tGata1(Zf)\t-0.230130\t-\n",
      "chr1\t53345\t53545\tGata1(Zf)\t-1.404719\t+\n",
      "chr1\t57025\t57225\tGata1(Zf)\t0.037453\t+\n"
     ]
    }
   ],
   "source": [
    "! head {slopped_cluster_deduplicated_scored_motif_scans_filepath}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d8226eb-2dcf-49fb-9bc5-47f188e023f0",
   "metadata": {},
   "source": [
    "# Preview BED file of scored intervals "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "83a6f8d0-c0ca-4d42-96c0-ac62ecd52f38",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>22919</td>\n",
       "      <td>23119</td>\n",
       "      <td>Gata1(Zf)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>27984</td>\n",
       "      <td>28184</td>\n",
       "      <td>Gata1(Zf)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>32492</td>\n",
       "      <td>32692</td>\n",
       "      <td>Gata1(Zf)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>32657</td>\n",
       "      <td>32857</td>\n",
       "      <td>Gata1(Zf)</td>\n",
       "      <td>-0.839098</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42233</td>\n",
       "      <td>42433</td>\n",
       "      <td>Gata1(Zf)</td>\n",
       "      <td>-0.983107</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>469038</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57084568</td>\n",
       "      <td>57084768</td>\n",
       "      <td>Gata1(Zf)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>469039</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57128542</td>\n",
       "      <td>57128742</td>\n",
       "      <td>Gata1(Zf)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>469040</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57138705</td>\n",
       "      <td>57138905</td>\n",
       "      <td>Gata1(Zf)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>469041</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57155332</td>\n",
       "      <td>57155532</td>\n",
       "      <td>Gata1(Zf)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>469042</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57192874</td>\n",
       "      <td>57193074</td>\n",
       "      <td>Gata1(Zf)</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>469043 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chr     Start       End       Name     Score Strand\n",
       "0       chr1     22919     23119  Gata1(Zf)  0.000000      +\n",
       "1       chr1     27984     28184  Gata1(Zf)  0.000000      -\n",
       "2       chr1     32492     32692  Gata1(Zf)  0.000000      +\n",
       "3       chr1     32657     32857  Gata1(Zf) -0.839098      -\n",
       "4       chr1     42233     42433  Gata1(Zf) -0.983107      +\n",
       "...      ...       ...       ...        ...       ...    ...\n",
       "469038  chrY  57084568  57084768  Gata1(Zf)  0.000000      +\n",
       "469039  chrY  57128542  57128742  Gata1(Zf)  0.000000      -\n",
       "469040  chrY  57138705  57138905  Gata1(Zf)  0.000000      +\n",
       "469041  chrY  57155332  57155532  Gata1(Zf)  0.000000      -\n",
       "469042  chrY  57192874  57193074  Gata1(Zf)  0.000000      +\n",
       "\n",
       "[469043 rows x 6 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed_columns = 'Chr Start End Name Score Strand'.split()\n",
    "bed_df = pd.read_csv(cluster_deduplicated_scored_motif_scans_filepath, sep = '\\t', header = None, names = bed_columns)\n",
    "bed_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af2b7dc9-7e1b-48ae-8fb6-0de31179f623",
   "metadata": {},
   "source": [
    "# Visualize score distribution\n",
    "Like most tools, MEPP prefers normal score distributions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "b9521768-912a-4c41-94fe-02e3fda7cfa8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<AxesSubplot:title={'center':'Score'}>]], dtype=object)"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYQAAAEICAYAAABfz4NwAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAYuElEQVR4nO3dfZBddZ3n8ffHRDElgiDSiyQ7wSU+ABkfiMiU41Q7ccY4OMAfsBvXkTCb2exSjINVmdLg1NY+VFEFs+ug1gzUZsEhoDOQRR2yKo5smN6HGQiCo0aexoxEiGRA5EHiDmjjd/+4v9ZL0+m+3en0vRfer6pbfe7vnt8539Pp9Of+fufc06kqJEl6Ub8LkCQNBgNBkgQYCJKkxkCQJAEGgiSpMRAkSYCBIElqDARpP5L8cpK/SfJEkkeT/HWSt/a7LulgWdzvAqRBlOQw4AvAecBW4CXAO4Cn53Efi6rqmfnannSgHCFIU3stQFX9eVU9U1X/WFVfqapvAiT510nuTvJkkruSvKW1vyHJWJLHk9yZ5PSJDSa5KsnlSb6U5EfAO5O8Oslnk3w/yX1Jfq8vRythIEj783fAM0m2JHlPkiMmXkhyNvAfgHOAw4DTgR8keTHwP4CvAEcDHwQ+k+R1Xdv9l8BFwMuBv2nrfwM4FlgNfCjJuw/ysUlTMhCkKVTVD4FfBgr4b8D3k2xLMgL8DvCHVfXV6thVVd8FTgUOBS6uqh9X1c10pp3e17XpG6rqr6vqp8BK4FVV9Z/a+t9p+1q7cEcq/ZznEKT9qKq7gXMBkrwe+DTwcWAZ8PdTdHk18ED7ZT/hu3Te/U94oGv5F4BXJ3m8q20R8H8OsHRpTgwEqQdVdU+Sq4B/Q+eX+j+bYrUHgWVJXtQVCv+UzvTTzzbVtfwAcF9VrTgIJUuz5pSRNIUkr0+yMcnS9nwZnamfW4ErgN9PcnI6jk/yC8AO4EfAh5O8OMko8JvAtfvZzW3AD5N8JMmSJIuSnOSlreoXA0Ga2pPA24Ad7YqgW4FvARur6r/TOTH8Z229vwCOrKof0znB/B7gEeAy4JyqumeqHbRLTn8TeBNwX+tzBXD4QTsqaRrxD+RIksARgiSpMRAkSYCBIElqDARJEjDEn0M46qijavny5Qu6zx/96Ee87GUvW9B9zodhrRuGt/ZhrRuGt/ZhrRsWtvY77rjjkap61VSvDW0gLF++nNtvv31B9zk2Nsbo6OiC7nM+DGvdMLy1D2vdMLy1D2vdsLC1J/nu/l5zykiSBPQYCElekeT6JPe0W/7+UpIjk9yU5Nvta/fdIC9MsivJvd13bmyf7NzZXvtkkrT2Q5Jc19p3JFk+70cqSZpWryOETwBfrqrXA28E7gY2AdvbfVi2t+ckOYHO3RpPBNYAlyVZ1LZzObABWNEea1r7euCxqjoeuBS45ACPS5I0SzMGQvvLUb8CXAnQbtP7OHAGsKWttgU4sy2fAVxbVU9X1X3ALuCUJMcAh1XVLdX5ePTVk/pMbOt6YPXE6EGStDB6Oan8GuD7wJ8meSNwB3ABMFJVewGqam+So9v6x9K578uEPa3tJ215cvtEnwfatsaTPAG8ks69XX4myQY6IwxGRkYYGxvr7Sjnyb59+xZ8n/NhWOuG4a19WOuG4a19WOuGwam9l0BYDLwF+GBV7UjyCdr00H5M9c6+pmmfrs+zG6o2A5sBVq1aVQt9RcGwXsUwrHXD8NY+rHXD8NY+rHXD4NTeyzmEPcCeqtrRnl9PJyAeatNAtK8Pd62/rKv/Ujr3id/Tlie3P6tPksV07vb46GwPRpI0dzMGQlX9A/BA19+FXQ3cBWwD1rW2dcANbXkbsLZdOXQcnZPHt7XppSeTnNrOD5wzqc/Ets4Cbi5vwypJC6rXD6ZN/LHwlwDfAX6bTphsTbIeuB84G6Cq7kyylU5ojAPnt/u+A5wHXAUsAW5sD+icsL4myS46IwP/pqwkLbCeAqGqvg6smuKl1ftZ/yI6f0BkcvvtwElTtD9FCxRp2C3f9EU2rhzn3E1fBGD3xaf1uSKpN35SWZIEGAiSpMZAkCQBBoIkqTEQJEmAgSBJagwESRJgIEiSGgNBkgQYCJKkxkCQJAEGgiSpMRAkSYCBIElqDARJEmAgSJIaA0GSBBgIkqTGQJAkAQaCJKkxECRJgIEgSWoMBEkSYCBIkhoDQZIEGAiSpKanQEiyO8nOJF9PcntrOzLJTUm+3b4e0bX+hUl2Jbk3ybu72k9u29mV5JNJ0toPSXJda9+RZPk8H6ckaQazGSG8s6reVFWr2vNNwPaqWgFsb89JcgKwFjgRWANclmRR63M5sAFY0R5rWvt64LGqOh64FLhk7ockSZqLA5kyOgPY0pa3AGd2tV9bVU9X1X3ALuCUJMcAh1XVLVVVwNWT+kxs63pg9cToQZK0MNL53TzDSsl9wGNAAf+1qjYnebyqXtG1zmNVdUSSPwZurapPt/YrgRuB3cDFVfWu1v4O4CNV9d4k3wLWVNWe9trfA2+rqkcm1bGBzgiDkZGRk6+99toDO/pZ2rdvH4ceeuiC7nM+DGvdMJy17/zeE4wsgYf+sfN85bGH97egWRrG7zkMb92wsLW/853vvKNrpudZFve4jbdX1YNJjgZuSnLPNOtO9c6+pmmfrs+zG6o2A5sBVq1aVaOjo9MWPd/GxsZY6H3Oh2GtG4az9nM3fZGNK8f52M7Of6/d7x/tb0GzNIzfcxjeumFwau9pyqiqHmxfHwY+D5wCPNSmgWhfH26r7wGWdXVfCjzY2pdO0f6sPkkWA4cDj87+cCRJczVjICR5WZKXTywDvw58C9gGrGurrQNuaMvbgLXtyqHj6Jw8vq2q9gJPJjm1nR84Z1KfiW2dBdxcvcxlSZLmTS9TRiPA59s53sXAn1XVl5N8FdiaZD1wP3A2QFXdmWQrcBcwDpxfVc+0bZ0HXAUsoXNe4cbWfiVwTZJddEYGa+fh2CRJszBjIFTVd4A3TtH+A2D1fvpcBFw0RfvtwElTtD9FCxRJUn/4SWVJEmAgSJIaA0GSBBgIkqTGQJAkAQaCJKkxECRJgIEgSWoMBEkSYCBIkhoDQZIEGAiSpMZAkCQBBoIkqTEQJEmAgSBJagwESRJgIEiSGgNBkgQYCJKkxkCQJAEGgiSpMRAkSYCBIElqDARJEmAgSJIaA0GSBMwiEJIsSvK3Sb7Qnh+Z5KYk325fj+ha98Iku5Lcm+TdXe0nJ9nZXvtkkrT2Q5Jc19p3JFk+j8coSerBbEYIFwB3dz3fBGyvqhXA9vacJCcAa4ETgTXAZUkWtT6XAxuAFe2xprWvBx6rquOBS4FL5nQ0kqQ56ykQkiwFTgOu6Go+A9jSlrcAZ3a1X1tVT1fVfcAu4JQkxwCHVdUtVVXA1ZP6TGzremD1xOhBkrQwFve43seBDwMv72obqaq9AFW1N8nRrf1Y4Nau9fa0tp+05cntE30eaNsaT/IE8Ergke4ikmygM8JgZGSEsbGxHsufH/v27Vvwfc6HYa0bhrP2jSvHGVnS+QoMXf3D+D2H4a0bBqf2GQMhyXuBh6vqjiSjPWxzqnf2NU37dH2e3VC1GdgMsGrVqhod7aWc+TM2NsZC73M+DGvdMJy1n7vpi2xcOc7Hdnb+e+1+/2h/C5qlYfyew/DWDYNTey8jhLcDpyf5DeClwGFJPg08lOSYNjo4Bni4rb8HWNbVfynwYGtfOkV7d589SRYDhwOPzvGYJElzMOM5hKq6sKqWVtVyOieLb66q3wK2AevaauuAG9ryNmBtu3LoODonj29r00tPJjm1nR84Z1KfiW2d1fbxnBGCJOng6fUcwlQuBrYmWQ/cD5wNUFV3JtkK3AWMA+dX1TOtz3nAVcAS4Mb2ALgSuCbJLjojg7UHUJckaQ5mFQhVNQaMteUfAKv3s95FwEVTtN8OnDRF+1O0QJEk9YefVJYkAQaCJKkxECRJgIEgSWoMBEkSYCBIkhoDQZIEGAiSpMZAkCQBBoIkqTmQexlJapZv+mK/S5AOmCMESRJgIEiSGqeMpDlymkjPN44QJEmAgSBJapwykhZQ9zTT7otP62Ml0nMZCNIseN5Az2cGgnSQGSIaFp5DkCQBBoIkqTEQJEmAgSBJagwESRJgIEiSGgNBkgT0EAhJXprktiTfSHJnkv/Y2o9MclOSb7evR3T1uTDJriT3Jnl3V/vJSXa21z6ZJK39kCTXtfYdSZYfhGOVJE2jlxHC08CvVtUbgTcBa5KcCmwCtlfVCmB7e06SE4C1wInAGuCyJIvati4HNgAr2mNNa18PPFZVxwOXApcc+KFJkmZjxkCojn3t6Yvbo4AzgC2tfQtwZls+A7i2qp6uqvuAXcApSY4BDquqW6qqgKsn9ZnY1vXA6onRgyRpYfR064r2Dv8O4HjgT6pqR5KRqtoLUFV7kxzdVj8WuLWr+57W9pO2PLl9os8DbVvjSZ4AXgk8MqmODXRGGIyMjDA2NtbjYc6Pffv2Lfg+58Ow1g2DV/vGleM9rTeyZOZ1B+m4ug3a97xXw1o3DE7tPQVCVT0DvCnJK4DPJzlpmtWnemdf07RP12dyHZuBzQCrVq2q0dHRacqYf2NjYyz0PufDsNYNg1f7uT3el2jjynE+tnP6/1673z86DxXNv0H7nvdqWOuGwal9VlcZVdXjwBiduf+H2jQQ7evDbbU9wLKubkuBB1v70inan9UnyWLgcODR2dQmSTowvVxl9Ko2MiDJEuBdwD3ANmBdW20dcENb3gasbVcOHUfn5PFtbXrpySSntvMD50zqM7Gts4Cb23kGSdIC6WXK6BhgSzuP8CJga1V9IcktwNYk64H7gbMBqurOJFuBu4Bx4Pw25QRwHnAVsAS4sT0ArgSuSbKLzshg7XwcnCSpdzMGQlV9E3jzFO0/AFbvp89FwEVTtN8OPOf8Q1U9RQsUSVJ/+EllSRJgIEiSGgNBkgT4N5WlGfk3kfVC4QhBkgQYCJKkxkCQJAEGgiSpMRAkSYCBIElqvOxU6pPuy1l3X3xaHyuROhwhSJIAA0GS1BgIkiTAQJAkNQaCJAkwECRJjYEgSQIMBElSYyBIkgA/qSxNyT+KoxciRwiSJMBAkCQ1ThlJA8Ab3WkQOEKQJAEGgiSpmTEQkixL8ldJ7k5yZ5ILWvuRSW5K8u329YiuPhcm2ZXk3iTv7mo/OcnO9tonk6S1H5Lkuta+I8nyg3CskqRp9DJCGAc2VtUbgFOB85OcAGwCtlfVCmB7e057bS1wIrAGuCzJoraty4ENwIr2WNPa1wOPVdXxwKXAJfNwbJKkWZgxEKpqb1V9rS0/CdwNHAucAWxpq20BzmzLZwDXVtXTVXUfsAs4JckxwGFVdUtVFXD1pD4T27oeWD0xepAkLYxZXWXUpnLeDOwARqpqL3RCI8nRbbVjgVu7uu1pbT9py5PbJ/o80LY1nuQJ4JXAI5P2v4HOCIORkRHGxsZmU/4B27dv34Lvcz4Ma93Qv9o3rhw/oP4jS+a+jX7/Ww3rz8uw1g2DU3vPgZDkUOCzwIeq6ofTvIGf6oWapn26Ps9uqNoMbAZYtWpVjY6OzlD1/BobG2Oh9zkfhrVu6F/t5x7gJ5U3rhznYzvndlX37vePHtC+D9Sw/rwMa90wOLX3dJVRkhfTCYPPVNXnWvNDbRqI9vXh1r4HWNbVfSnwYGtfOkX7s/okWQwcDjw624ORJM1dL1cZBbgSuLuq/qjrpW3Aura8Drihq31tu3LoODonj29r00tPJjm1bfOcSX0mtnUWcHM7zyBJWiC9jGnfDnwA2Jnk663to8DFwNYk64H7gbMBqurOJFuBu+hcoXR+VT3T+p0HXAUsAW5sD+gEzjVJdtEZGaw9sMOSJM3WjIFQVf+Xqef4AVbvp89FwEVTtN8OnDRF+1O0QJEk9YefVJYkAd7cTvoZ/waCXugcIUiSAANBktQYCJIkwECQJDUGgiQJMBAkSY2BIEkC/ByCNHC6Pw+x++LT+liJXmgcIUiSAANBktQYCJIkwECQJDUGgiQJ8CojvcB5h1Pp5xwhSJIAA0GS1BgIkiTAQJAkNQaCJAkwECRJjZedSgPMG91pITlCkCQBBoIkqTEQJElAD+cQknwKeC/wcFWd1NqOBK4DlgO7gX9eVY+11y4E1gPPAL9XVX/Z2k8GrgKWAF8CLqiqSnIIcDVwMvAD4F9U1e55O0JpEm9XIU2tlxHCVcCaSW2bgO1VtQLY3p6T5ARgLXBi63NZkkWtz+XABmBFe0xscz3wWFUdD1wKXDLXg5Ekzd2MgVBV/xt4dFLzGcCWtrwFOLOr/dqqerqq7gN2AackOQY4rKpuqaqiMyI4c4ptXQ+sTpK5HY4kaa7meg5hpKr2ArSvR7f2Y4EHutbb09qObcuT25/Vp6rGgSeAV86xLknSHM335xCmemdf07RP1+e5G0820Jl2YmRkhLGxsTmUOHf79u1b8H3Oh2GtGw5O7RtXjs/r9qYysmT+97NQ/4bD+vMyrHXD4NQ+10B4KMkxVbW3TQc93Nr3AMu61lsKPNjal07R3t1nT5LFwOE8d4oKgKraDGwGWLVqVY2Ojs6x/LkZGxtjofc5H4a1bjg4tZ+7ACeVN64c52M75/f91u73j87r9vZnWH9ehrVuGJza5zpltA1Y15bXATd0ta9NckiS4+icPL6tTSs9meTUdn7gnEl9JrZ1FnBzO88gSVpAvVx2+ufAKHBUkj3AvwcuBrYmWQ/cD5wNUFV3JtkK3AWMA+dX1TNtU+fx88tOb2wPgCuBa5LsojMyWDsvRyZJmpUZA6Gq3refl1bvZ/2LgIumaL8dOGmK9qdogSJJ6h9vbicNCW90p4PNQNALgp9OlmbmvYwkSYCBIElqDARJEmAgSJIaA0GSBHiVkZ7HvLJImh0DQRpCfiZBB4NTRpIkwECQJDUGgiQJMBAkSY0nlfW84pVF0tw5QpAkAY4QpKHnJaiaL44QJEmAIwQ9D3jeQJofjhAkSYAjBOl5xfMJOhAGgoaS00TS/HPKSJIEOELQEHFUMDtOH2m2HCFIkgBHCBpwO7/3BOc6MpAWhIGggdM91bFxZR8LeR5x+ki9MBDUN54T6A/DQfszMIGQZA3wCWARcEVVXdznkjRP/MU/uKb6t9m4cpzRhS9FA2AgAiHJIuBPgF8D9gBfTbKtqu7qb2WazF/uLwy9/Ds7unj+GYhAAE4BdlXVdwCSXAucAbwgA2G+f+luXDnuiVnNu0F5c2AwzZ9UVb9rIMlZwJqq+p32/APA26rqdyettwHY0J6+Drh3QQuFo4BHFnif82FY64bhrX1Y64bhrX1Y64aFrf0XqupVU70wKCOETNH2nKSqqs3A5oNfztSS3F5Vq/q1/7ka1rpheGsf1rpheGsf1rphcGoflA+m7QGWdT1fCjzYp1ok6QVpUALhq8CKJMcleQmwFtjW55ok6QVlIKaMqmo8ye8Cf0nnstNPVdWdfS5rKn2brjpAw1o3DG/tw1o3DG/tw1o3DEjtA3FSWZLUf4MyZSRJ6jMDQZIEGAizluSDSe5NcmeSP+x3PbOV5PeTVJKj+l1LL5L85yT3JPlmks8neUW/a5pJkjXtZ2RXkk39rqcXSZYl+askd7ef7Qv6XdNsJVmU5G+TfKHftfQqySuSXN9+xu9O8kv9rMdAmIUk76TzCepfrKoTgf/S55JmJckyOrcHub/ftczCTcBJVfWLwN8BF/a5nml13YblPcAJwPuSnNDfqnoyDmysqjcApwLnD0nd3S4A7u53EbP0CeDLVfV64I30uX4DYXbOAy6uqqcBqurhPtczW5cCH2aKD/0Nqqr6SlWNt6e30vmMyiD72W1YqurHwMRtWAZaVe2tqq+15Sfp/GI6tr9V9S7JUuA04Ip+19KrJIcBvwJcCVBVP66qx/tZk4EwO68F3pFkR5L/leSt/S6oV0lOB75XVd/ody0H4F8BN/a7iBkcCzzQ9XwPQ/SLFSDJcuDNwI4+lzIbH6fzZuenfa5jNl4DfB/40zbVdUWSl/WzoIH4HMIgSfI/gX8yxUt/QOf7dQSdIfVbga1JXlMDcu3uDLV/FPj1ha2oN9PVXVU3tHX+gM60xmcWsrY56Ok2LIMqyaHAZ4EPVdUP+11PL5K8F3i4qu5IMtrncmZjMfAW4INVtSPJJ4BNwL/rZ0HqUlXv2t9rSc4DPtcC4LYkP6VzU6rvL1R909lf7UlWAscB30gCnWmXryU5par+YQFLnNJ033OAJOuA9wKrByV8pzG0t2FJ8mI6YfCZqvpcv+uZhbcDpyf5DeClwGFJPl1Vv9XnumayB9hTVRMjsevpBELfOGU0O38B/CpAktcCL2EI7q5YVTur6uiqWl5Vy+n8IL5lEMJgJu0PJ30EOL2q/l+/6+nBUN6GJZ13ClcCd1fVH/W7ntmoqguramn72V4L3DwEYUD7//dAkte1ptX0+Zb/jhBm51PAp5J8C/gxsG4I3rEOuz8GDgFuaqObW6vq3/a3pP0botuwTPZ24APAziRfb20fraov9a+kF4QPAp9pbx6+A/x2P4vx1hWSJMApI0lSYyBIkgADQZLUGAiSJMBAkCQ1BoIkCTAQJEnN/wcxrFiSNe9NRwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "bed_df[['Score']].hist(bins=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca92baba-4dff-405b-82bb-3fd35ce7d1d2",
   "metadata": {},
   "source": [
    "# Download JASPAR-converted HOMER vertebrate motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "52ca7af9-005b-47ca-b479-3e9242d1061a",
   "metadata": {},
   "outputs": [],
   "source": [
    "motifs_url = 'https://raw.githubusercontent.com/npdeloss/mepp/main/data/homer.motifs.txt'\n",
    "motifs_filepath = 'homer.motifs.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "feaeb58e-ea36-4e47-8176-a4848be42025",
   "metadata": {},
   "outputs": [],
   "source": [
    "wget_motifs_cmd = (\n",
    "    f'wget -nc -O {motifs_filepath} \"{motifs_url}\"'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "a148e7d8-b81e-47c7-846c-800c3df2f702",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wget -nc -O homer.motifs.txt \"https://raw.githubusercontent.com/npdeloss/mepp/main/data/homer.motifs.txt\"\n",
      "--2022-01-27 12:16:23--  https://raw.githubusercontent.com/npdeloss/mepp/main/data/homer.motifs.txt\n",
      "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.108.133, ...\n",
      "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 156010 (152K) [text/plain]\n",
      "Saving to: ‘homer.motifs.txt’\n",
      "\n",
      "homer.motifs.txt    100%[===================>] 152.35K  --.-KB/s    in 0.006s  \n",
      "\n",
      "2022-01-27 12:16:23 (25.2 MB/s) - ‘homer.motifs.txt’ saved [156010/156010]\n",
      "\n",
      "CPU times: user 7.09 ms, sys: 13.1 ms, total: 20.2 ms\n",
      "Wall time: 416 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "run_wget_motifs_cmds = run_cmd = True\n",
    "for cmd in [wget_motifs_cmd]:\n",
    "    print(cmd)\n",
    "    if run_cmd:\n",
    "        ! {cmd}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b226c499-a973-406e-a609-dd95f9bbe566",
   "metadata": {},
   "source": [
    "# Convert scored bed file to scored sequences, and run MEPP analysis\n",
    "In `mepp.get_scored_fasta` we handle reverse complementation of sequence according the BED interval's strand value. We can then pipe that directly into MEPP."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b28a806-acfe-480f-9bc7-b0c75a6cccb3",
   "metadata": {},
   "source": [
    "## Explanation of parameters\n",
    "\n",
    "`python -m mepp.get_scored_fasta`\n",
    "* Utility for extracting scored FASTA files (sequence score in header) from scored bed files\n",
    "    * `-fi {genome_fa_filepath}`: Extract sequence from the specified genome FASTA (required).\n",
    "    * `-bed {bed_filepath}`: Extract sequences from the intervals specified in this BED file (required). \n",
    "    \n",
    "`|python -m mepp.cli`\n",
    "* Pipe output from previous command into MEPP\n",
    "    * `--fa - `: Receive scored FASTA from the output of the previous command (Required).\n",
    "    * `--motifs {motifs_filepath}:` Analyze these motifs from a JASPAR-formatted motif matrix collection file (Required).\n",
    "    * `--out {mepp_filepath} `: Output to this directory (Required).\n",
    "    * `--perms 100 `: Use 100 permutations for confidence interval statistics (Default: 1000, can be costly in time & memory).\n",
    "    * `--batch 1000`: Use 1000 as tensorflow batch size (Default: 1000, adjust according to machine memory)\n",
    "    * `dgt 50`: Only analyze sequences with less than 50% degenerate base content (Default: 100, adjust according to analysis needs)\n",
    "    * `--jobs 20`: Use 20 jobs for multithreaded tasks. (Default: Use all cores)\n",
    "    * `--gjobs 20`: Use 20 jobs for multithreaded tasks optimizable by Tensorflow GPU usage. (Default: 1)\n",
    "    * `--nogpu`: Don't use the GPU (Default: Use GPU.)\n",
    "        * if set, `--gjobs` is simply the number of cores used to process motifs in parallel.\n",
    "    * `--dpi 100`: DPI of plots. Important, since the motif occurrence heatmap is DPI-dependent. (Default: 300)\n",
    "    * `--orientations +,- `: Analyze these orientations of the motifs (Forward, and reverse). (Default: +,+/-, analyze Forward, and non-orientation specific)\n",
    "    * Not specified here:\n",
    "        * ` --margin {INTEGER}`: Number of bases along either side of motif to \"blur\" motif matches for smoothing. (Default: 2)\n",
    "            * It can be useful to set this depending on how strictly your sequences have been centered. If centering on ChIP-seq peak centers, consider a larger margin."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "d15aa209-264e-43e2-aace-c41ead622b93",
   "metadata": {},
   "outputs": [],
   "source": [
    "mepp_filepath = slopped_cluster_deduplicated_scored_motif_scans_filepath[:-len('.bed')]+f'.mepp'\n",
    "\n",
    "mepp_cmd = (\n",
    "    f'python -m mepp.get_scored_fasta -fi {genome_fa_filepath} '\n",
    "    f'-bed {slopped_cluster_deduplicated_scored_motif_scans_filepath} '\n",
    "    f'|python -m mepp.cli '\n",
    "    f'--fa - '\n",
    "    f'--motifs {motifs_filepath} '\n",
    "    f'--out {mepp_filepath} '\n",
    "    f'--perms 100 '\n",
    "    f'--batch 1000 '\n",
    "    f'--dgt 50 '\n",
    "    f'--jobs 15 '\n",
    "    f'--gjobs 15 '\n",
    "    f'--nogpu '\n",
    "    f'--dpi 100 '\n",
    "    f'--orientations +,- '\n",
    "    f'&> {mepp_filepath}.log'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "078a262a-b967-4ce4-94eb-2c6bb9dd1c0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "run_mepp_cmd = run_cmd = True\n",
    "for cmd in [mepp_cmd]:\n",
    "    print(cmd)\n",
    "    if run_cmd:\n",
    "        ! {cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2385424d-cb7b-4698-a7b0-44034140bbb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "! tail {mepp_filepath}.log"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c134deb-b20c-4902-9592-cd669c3d6e02",
   "metadata": {},
   "source": [
    "# Show links to MEPP HTML outputs\n",
    "MEPP outputs HTML files that are useful for visualizing and navigating your data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "8e8c4ed5-2b5c-4286-9d9c-694dbf51c729",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import display, Markdown"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "97b6c767-f456-4775-9841-e7d9d31a704a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "mepp_results_table_fwd_md = f'[Results table, + orientation]({mepp_filepath}/results_table_orientation_fwd.html)'\n",
    "mepp_clustermap_fwd_md = f'[Clustermap, + orientation]({mepp_filepath}/clustermap_orientation_fwd.html)'\n",
    "\n",
    "mepp_results_table_rev_md = f'[Results table, - orientation]({mepp_filepath}/results_table_orientation_rev.html)'\n",
    "mepp_clustermap_rev_md = f'[Clustermap, - orientation]({mepp_filepath}/clustermap_orientation_rev.html)'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "09de2913-e7c7-4ad3-aafe-a4287b7778e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "[Results table, + orientation](gata.scans.scored_by.hct116.vs.k562.log2fc.bw.cluster_deduplicated.slop_0.mepp/results_table_orientation_fwd.html)"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "[Clustermap, + orientation](gata.scans.scored_by.hct116.vs.k562.log2fc.bw.cluster_deduplicated.slop_0.mepp/clustermap_orientation_fwd.html)"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "[Results table, - orientation](gata.scans.scored_by.hct116.vs.k562.log2fc.bw.cluster_deduplicated.slop_0.mepp/results_table_orientation_rev.html)"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/markdown": [
       "[Clustermap, - orientation](gata.scans.scored_by.hct116.vs.k562.log2fc.bw.cluster_deduplicated.slop_0.mepp/clustermap_orientation_rev.html)"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(Markdown(mepp_results_table_fwd_md))\n",
    "display(Markdown(mepp_clustermap_fwd_md))\n",
    "display(Markdown(mepp_results_table_rev_md))\n",
    "display(Markdown(mepp_clustermap_rev_md))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3520bfd9-dad1-4c1e-bd21-ae1782d4616e",
   "metadata": {},
   "source": [
    "# Example commands for set-based analysis with CentriMo\n",
    "Threshold top and bottom 10% of scored sequences, then use these as positive/negative inputs to set-based MEA, e.g. CentriMo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "06ca5cd4-a2ef-4b7e-863c-0002b9b6f2ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "percent = 10\n",
    "lower_percent = percent\n",
    "higher_percent = 100.0-percent\n",
    "lower_thresh, upper_thresh = list(np.percentile(bed_df['Score'], [lower_percent, higher_percent]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "3c3d05bf-97ed-49fc-97ad-3fb8231d7819",
   "metadata": {},
   "outputs": [],
   "source": [
    "upper_slopped_cluster_deduplicated_scored_motif_scans_filepath = cluster_deduplicated_scored_motif_scans_filepath[:-len('.bed')]+'.upper.bed'\n",
    "lower_slopped_cluster_deduplicated_scored_motif_scans_filepath = cluster_deduplicated_scored_motif_scans_filepath[:-len('.bed')]+'.lower.bed'\n",
    "\n",
    "upper_slopped_cluster_deduplicated_scored_motif_scans_fa_filepath = cluster_deduplicated_scored_motif_scans_filepath[:-len('.bed')]+'.upper.fa'\n",
    "lower_slopped_cluster_deduplicated_scored_motif_scans_fa_filepath = cluster_deduplicated_scored_motif_scans_filepath[:-len('.bed')]+'.lower.fa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "6ba03c8f-024f-4ab4-980d-d23f099617b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "upper_bed_df = bed_df[bed_df['Score']>=upper_thresh].copy()\n",
    "lower_bed_df = bed_df[bed_df['Score']<=lower_thresh].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "1dc22f20-6358-47a9-957b-6c19ac807024",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t50012\t50212\tGata1(Zf)\t1.275901\t-\n",
      "chr1\t1935697\t1935897\tGata1(Zf)\t0.923212\t-\n",
      "chr1\t2003160\t2003360\tGata1(Zf)\t0.865649\t+\n",
      "chr1\t2409694\t2409894\tGata1(Zf)\t1.178493\t-\n",
      "chr1\t2752677\t2752877\tGata1(Zf)\t1.020837\t+\n",
      "chr1\t2768240\t2768440\tGata1(Zf)\t1.618224\t-\n",
      "chr1\t2787196\t2787396\tGata1(Zf)\t0.883021\t+\n",
      "chr1\t2975991\t2976191\tGata1(Zf)\t1.141857\t+\n",
      "chr1\t3057763\t3057963\tGata1(Zf)\t0.917672\t-\n",
      "chr1\t3068900\t3069100\tGata1(Zf)\t4.611864\t+\n"
     ]
    }
   ],
   "source": [
    "upper_bed_df.to_csv(upper_slopped_cluster_deduplicated_scored_motif_scans_filepath, sep = '\\t', index = False, header = None)\n",
    "! head {upper_slopped_cluster_deduplicated_scored_motif_scans_filepath}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "e1a6e51d-e799-4026-ac52-8bf15e3a02cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t53345\t53545\tGata1(Zf)\t-1.404719\t+\n",
      "chr1\t794774\t794974\tGata1(Zf)\t-1.266198\t+\n",
      "chr1\t796048\t796248\tGata1(Zf)\t-2.179768\t+\n",
      "chr1\t796306\t796506\tGata1(Zf)\t-1.406486\t+\n",
      "chr1\t803941\t804141\tGata1(Zf)\t-1.033328\t+\n",
      "chr1\t810799\t810999\tGata1(Zf)\t-1.12269\t+\n",
      "chr1\t831715\t831915\tGata1(Zf)\t-2.751448\t-\n",
      "chr1\t842880\t843080\tGata1(Zf)\t-4.897698\t+\n",
      "chr1\t843543\t843743\tGata1(Zf)\t-1.04601\t+\n",
      "chr1\t848540\t848740\tGata1(Zf)\t-1.029326\t+\n"
     ]
    }
   ],
   "source": [
    "lower_bed_df.to_csv(lower_slopped_cluster_deduplicated_scored_motif_scans_filepath, sep = '\\t', index = False, header = None)\n",
    "! head {lower_slopped_cluster_deduplicated_scored_motif_scans_filepath}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "736ac51c-043a-4a75-bf19-3511e3479617",
   "metadata": {},
   "outputs": [],
   "source": [
    "upper_fa_cmd = ( f'python -m mepp.get_scored_fasta -fi {genome_fa_filepath} '\n",
    "    f'-bed {upper_slopped_cluster_deduplicated_scored_motif_scans_filepath} '\n",
    "    f'> {upper_slopped_cluster_deduplicated_scored_motif_scans_fa_filepath}'\n",
    ")\n",
    "\n",
    "lower_fa_cmd = ( f'python -m mepp.get_scored_fasta -fi {genome_fa_filepath} '\n",
    "    f'-bed {lower_slopped_cluster_deduplicated_scored_motif_scans_filepath} '\n",
    "    f'> {lower_slopped_cluster_deduplicated_scored_motif_scans_fa_filepath}'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "f7ad5710-326b-43a8-b5d8-8406aedcb4e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python -m mepp.get_scored_fasta -fi hg38.fa -bed gata.scans.scored_by.hct116.vs.k562.log2fc.bw.cluster_deduplicated.upper.bed > gata.scans.scored_by.hct116.vs.k562.log2fc.bw.cluster_deduplicated.upper.fa\n",
      "/home/provnpd0/miniconda3/envs/mepp_walkthrough/lib/python3.9/runpy.py:127: RuntimeWarning: 'mepp.get_scored_fasta' found in sys.modules after import of package 'mepp', but prior to execution of 'mepp.get_scored_fasta'; this may result in unpredictable behaviour\n",
      "  warn(RuntimeWarning(msg))\n",
      "python -m mepp.get_scored_fasta -fi hg38.fa -bed gata.scans.scored_by.hct116.vs.k562.log2fc.bw.cluster_deduplicated.lower.bed > gata.scans.scored_by.hct116.vs.k562.log2fc.bw.cluster_deduplicated.lower.fa\n",
      "/home/provnpd0/miniconda3/envs/mepp_walkthrough/lib/python3.9/runpy.py:127: RuntimeWarning: 'mepp.get_scored_fasta' found in sys.modules after import of package 'mepp', but prior to execution of 'mepp.get_scored_fasta'; this may result in unpredictable behaviour\n",
      "  warn(RuntimeWarning(msg))\n",
      "CPU times: user 841 ms, sys: 164 ms, total: 1 s\n",
      "Wall time: 1min 11s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "run_thresh_fa_cmds = run_cmd = True\n",
    "for cmd in [upper_fa_cmd, lower_fa_cmd]:\n",
    "    print(cmd)\n",
    "    if run_cmd:\n",
    "        ! {cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "afa2bedc-d1bb-4632-bde7-092a0b61cf3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(46905, 6)"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "upper_bed_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "7cbc6564-4194-4f13-b491-6f0cd11f85ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(46905, 6)"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lower_bed_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67b8c8dd-a706-4b3d-8262-8fe392afacaa",
   "metadata": {},
   "source": [
    "# Example equivalent Centrimo command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "52d98bb6-21e5-44b3-a42d-36b7ed4f84fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir -p gata.scans.scored_by.hct116.vs.k562.log2fc.bw.cluster_deduplicated.slop_0.upper_vs_lower.centrimo ;$(which time) --verbose centrimo --oc gata.scans.scored_by.hct116.vs.k562.log2fc.bw.cluster_deduplicated.slop_0.upper_vs_lower.centrimo --neg gata.scans.scored_by.hct116.vs.k562.log2fc.bw.cluster_deduplicated.lower.fa --norc --sep --local --noseq gata.scans.scored_by.hct116.vs.k562.log2fc.bw.cluster_deduplicated.upper.fa homer.motifs.id_fixed.meme\n"
     ]
    }
   ],
   "source": [
    "meme_motifs_filepath = 'homer.motifs.id_fixed.meme'\n",
    "centrimo_filepath = mepp_filepath[:-len('.mepp')]+f'.upper_vs_lower.centrimo'\n",
    "centrimo_cmd = (\n",
    "    f'mkdir -p {centrimo_filepath} ;'\n",
    "    f'$(which time) --verbose '\n",
    "    f'centrimo --oc {centrimo_filepath} '\n",
    "    f'--neg {lower_slopped_cluster_deduplicated_scored_motif_scans_fa_filepath} '\n",
    "    f'--norc --sep --local --noseq '\n",
    "    f'{upper_slopped_cluster_deduplicated_scored_motif_scans_fa_filepath} '\n",
    "    f'{meme_motifs_filepath}'\n",
    ")\n",
    "print(centrimo_cmd)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:mepp_walkthrough]",
   "language": "python",
   "name": "conda-env-mepp_walkthrough-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
